var pathvisio={},pathvisioNS=pathvisioNS||{};pathvisioNS["tmp/pathvisio-js.html"]='<div id="pathvisio-js-container" style="position:relative; width:70%; height:auto; float:left;">\n  <div style="width:100%; height:100%; float:right; clear:both;" id="pathway-viewer">\n    <div style="width:80%; height:500px" id="pathway-container">\n<!--?xml version="1.0" encoding="UTF-8" standalone="no"?-->\n\n<!-- \nThis file serves as our template for standardized GPML pathway visual representations.\nThis file can be used for many purposes outside of pathvisio.js, including allowingar\nother projects to work better with GPML. For purposes of pathvisio.js, this file will\nbe the starting point for our JavaScript rendering of pathways. Every time we want to\nrender a GPML file on the browser, we will read a copy of this file into D3.js and\nmodify the copy by adding "use" statements, cloning markers, etc. in order to create\nthe desired pathway illustration in SVG.\n\nThe XML declaration above is taken from an example from the W3C \n<http://www.w3.org/TR/SVG/images/struct/use04.svg>,\nexcept I added:\nencoding="UTF-8" \nWe will want to change standalone to yes if we keep the CSS and\nJS all inside this document.\n\nDoctypes are not needed for SVG, and jwatt discourages their use:\nhttps://jwatt.org/svg/authoring/.\n\nStyle guides can be arbitrary, but for consistency of SVG markup for the pathvisio.js project,\n	I suggest using JS Watt\'s SVG authoring advice and Google\'s HTML and JavaScript Guides:\n	https://jwatt.org/svg/authoring/\n	http://google-styleguide.googlecode.com/svn/trunk/htmlcssguide.xml\n	http://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml\n\nJS Watt\'s advice is referenced from MDN:\nhttps://developer.mozilla.org/en-US/docs/Web/SVG\n\nFor testing, we will ensure this SVG file renders correctly with the following browsers and graphics programs:\nChrome (latest release)\n	Uses Skia graphics library\nFirefox (latest release)\n	Uses Azure graphics library\n	Safari (latest release)\n	Android Browser (latest release)\n	iOS Browser (latest release)\n	Internet Explorer (IE9 and subsequent versions)\nSquiggle SVG browser (latest release)\n	Uses Batik \n	Available for download at http://xmlgraphics.apache.org/batik/download.html\nSVG-Edit (latest release)\n	Version 2.6 (latest release as of 2013-07-08) available for download at \n	http://svg-edit.googlecode.com/svn/branches/2.6/editor/svg-editor.html\n	Inkscape\n	Uses livarot rendering engine but is in process of transitioning to Cairo\n	Available for download at http://inkscape.org/\n\n	Optional additional tests:\n	SvgWeb\n	Adobe Illustrator (CS6)\n		Either convert SVG to PDF and import PDF into Illustrator or convert SVG to .ai (Illustrator format)\n		with a converter like Uniconverter, available for download at\n		http://sk1project.org/modules.php?name=Products&product=uniconvertor\n		Opening this SVG directly with Illustrator does not work well.\n		In the future, we could use Uniconverter on the server to make it possible to download pathway images in\n		.ai (Illustrator) and .cdr (CorelDraw) formats.\n\n	For more information on SVG, these references are helpful:\n	[W3 Spec](http://www.w3.org/TR/SVG/expanded-toc.html)\n	[MDN on SVG](https://developer.mozilla.org/en-US/docs/Web/SVG)\n	-->\n\n	<svg id="pathway-image" version="1.1" baseProfile="full" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ev="http://www.w3.org/2001/xml-events" width="684.0" height="473.62184873949576" onmouseup="handleMouseUp(evt)" onmousedown="handleMouseDown(evt)" onmousemove="handleMouseMove(evt)">\n	<g>\n	<title>pathway defs for pathvisio.js</title>\n	<desc>\n	This SVG file contains all the graphical elements (markers and symbols in defs as well as\n	style data) used by the program pathvisio.js, which has two components: \n	1) a viewer for transforming GPML biological pathway data into an SVG visual representation and \n	2) an editor for creating both views and models for biological pathways.\n	</desc>\n	</g>\n\n	<defs>\n\n	<!-- ***************************\n	CSS Style Sheet\n	***************************\n\n	Note: pathvisio.js assumes the default element color is black.\n	-->\n\n	<style type="text/css">\n\n	svg {\n		color-interpolation: auto;\n		image-rendering: auto;\n		shape-rendering: auto;\n		vector-effect: non-scaling-stroke;\n		fill: white;\n	}\n\n	.drawing-board-color-fill {\n		fill: white;\n	}\n\n	.drawing-board-color-stroke {\n		stroke: white;\n	}\n\n	text {\n		font-family: Sans-Serif, Helvetica, Arial;\n		font-size: 10px;\n		fill: black;\n		fill-opacity: 1;\n		stroke: none;\n		text-anchor: middle;\n		font-size: 10px;\n		stroke: none;\n	}\n\n	.info-box {\n		font-family: Sans-Serif;\n		font-size: 10px;\n		fill: black;\n		stroke: none;\n		text-anchor: start;\n	}\n\n	.info-box-property-name {\n		font-weight: bold;\n	}\n\n	path.group {\n		fill-opacity: 0.098;\n		stroke: gray;\n		stroke-miterlimit: 1;\n		stroke-width: 1px;\n	}\n\n	path.group:hover {\n		fill-opacity: 0.2;\n		stroke-width: 1px;\n	}\n\n	path.group-none {\n		fill: rgb(180,180,100);\n		stroke-dasharray: 5,3;\n	}\n\n	path.group-group {\n		fill-opacity: 0;\n		stroke-width: 0;\n	}\n\n	path.group-complex {\n		fill: rgb(180,180,100);\n	}\n\n	path.group-pathway {\n		fill: lightgreen;\n		stroke-dasharray: 5,3;\n	}\n\n	use.data-node {\n		fill-opacity: 1;\n		stroke: black;\n		stroke-dasharray: 0;\n		stroke-miterlimit: 1;\n	}\n\n	use.gene-product {\n	}\n\n	use.metabolite {\n		stroke: blue;\n	}\n\n	text.metabolite {\n		fill: blue;\n	}\n\n	use.pathway {\n		stroke: rgb(20,150,30);\n		fill-opacity: 0;\n	}\n\n	text.pathway {\n		fill: rgb(20,150,30);\n		fill-opacity: 1;\n		stroke: none;\n	}\n\n	use.protein {\n	}\n\n	use.rna {\n	}\n\n	use.unknown {\n	}\n\n	use.label {\n		stroke: black;\n		stroke-width: 0;\n		fill-opacity: 0;\n		stroke-dasharray: 0;\n		stroke-miterlimit: 1;\n	}\n\n	use.shape {\n		fill-opacity: 0;\n		stroke: black;\n		stroke-dasharray: 0;\n		stroke-miterlimit: 1;\n	}\n\n	use.shape-none {\n		fill: none;\n		fill-opacity: 0;\n		stroke: none;\n	}\n\n	use.cellular-component {\n		fill-opacity: 0;\n		stroke: "#C0C0C0";\n	}\n\n	.graphical-line {\n		fill:none;\n		stroke: black; \n		stroke-width: 1px; \n	}\n\n	.interaction {\n		fill:none;\n		stroke: black; \n		stroke-width: 1px; \n	}\n\n	marker {\n		/* this is what should work per the spec\n		   stroke-dasharray: none; */\n		/* but I need to add this to make it work in Safari */\n		stroke-dasharray: 9999999999999999999999999;\n	}\n\n	.dashed-stroke {\n		stroke-dasharray: 5,3;\n	}\n</style>\n\n    <filter id="highlight" width="150%" height="150%">\n        <feOffset result="offOut" in="SourceGraphic" dx="30" dy="30"></feOffset>\n        <feGaussianBlur result="blurOut" in="offOut" stdDeviation="10"></feGaussianBlur>\n        <feBlend in="SourceGraphic" in2="blurOut" mode="normal"></feBlend>\n    </filter>\n\n	<!-- ***************************\n	Markers (Arrowheads) \n	*************************** -->\n\n	<!-- Here we generate a set of interaction markers for the default color (black). If we need other colors,\n	we need to clone the black marker and set the color for the clone to the desired color using d3.js.\n	I wish fill="currentColor" worked for markers, but that does not appear to be the case. -->\n\n	<!-- Each marker includes a small rectangle with a drawing-board-color-fill color to obscure the\n	ends of lines that might otherwise show up beneath the marker. Double lines require their own special\n	obscuring rects and are included as a double-line-hack-start/end marker, defined here and added in\n	gpml2json.js -->\n\n	<!-- arrow markers: triangular polygons, no stroke -->\n\n	<marker id="arrow-start-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="0" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="12,5 0,0 12,-5"></polygon>\n	</marker>\n\n	<marker id="arrow-end-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="-2" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="-12,5 0,0 -12,-5"></polygon>\n	</marker>\n\n\n	<!-- mim-conversion markers: triangular polygons, no stroke -->\n\n	<marker id="mim-conversion-start-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="0" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="11,5 0,0 11,-5"></polygon>\n	</marker>\n\n	<marker id="mim-conversion-end-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="-2" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="-11,5 0,0 -11,-5"></polygon>\n	</marker>\n\n	<!-- mim-stimulation markers: triangular polygons, drawing-board fill, black stroke -->\n\n	<marker id="mim-stimulation-start-black" class="drawing-board-color-fill" stroke="black" markerWidth="12" markerHeight="12" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="1" points="11,5 0,0 11,-5"></polygon>\n	</marker>\n\n	<marker id="mim-stimulation-end-black" class="drawing-board-color-fill" stroke="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-2" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="1" points="-11,5 0,0 -11,-5"></polygon>\n	</marker>\n\n	<!-- mim-necessary-stimulation markers: triangular polygons, drawing-board fill, black stroke; and vertical line -->\n\n	<marker id="mim-necessary-stimulation-start-black" class="drawing-board-color-fill" stroke="black" markerWidth="12" markerHeight="12" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 15 12">\n	<rect stroke="none" x="0" y="-0.6" width="2" height="1.2"></rect>\n	<line fill="none" stroke-width="1" x1="14" y1="-6" x2="14" y2="6"></line>\n	<polygon stroke-width="1" points="9,5 0,0 9,-5"></polygon>\n	</marker>\n\n	<marker id="mim-necessary-stimulation-end-black" class="drawing-board-color-fill" stroke="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="-15 -6 15 12">\n	<rect stroke="none" x="-2" y="-0.6" width="2" height="1.2"></rect>\n	<line fill="none" stroke-width="1" x1="-14" y1="-6" x2="-14" y2="6"></line>\n	<polygon stroke-width="1" points="-9,5 0,0 -9,-5"></polygon>\n	</marker>\n\n	<!-- t-bar markers: vertical line; and extended drawing-board rect -->\n\n	<marker id="t-bar-start-black" class="drawing-board-color-fill" stroke="black" markerWidth="16" markerHeight="16" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="5" height="1.2"></rect>\n	<line fill="none" stroke-width="1.6" x1="5" y1="-6" x2="5" y2="6"></line>\n	</marker>\n\n	<marker id="t-bar-end-black" class="drawing-board-color-fill" stroke="black" markerUnits="strokeWidth" markerWidth="16" markerHeight="16" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-5" y="-0.6" width="5" height="1.2"></rect>\n	<line fill="none" stroke-width="1.6" x1="-5" y1="-6" x2="-5" y2="6"></line>\n	</marker>\n	\n\n	<!-- mim-inhibition markers: vertical line; and extended drawing-board rect -->\n\n	<marker id="mim-inhibition-start-black" class="drawing-board-color-fill" stroke="black" markerWidth="16" markerHeight="16" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="5" height="1.2"></rect>\n	<line fill="none" stroke-width="1.6" x1="5" y1="-6" x2="5" y2="6"></line>\n	</marker>\n\n	<marker id="mim-inhibition-end-black" class="drawing-board-color-fill" stroke="black" markerUnits="strokeWidth" markerWidth="16" markerHeight="16" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-5" y="-0.6" width="5" height="1.2"></rect>\n	<line fill="none" stroke-width="1.6" x1="-5" y1="-6" x2="-5" y2="6"></line>\n	</marker>\n	\n	<!-- mim-binding markers: four-point polygon, no stroke -->\n\n	<marker id="mim-binding-start-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="0" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="12,6 0,0 12,-6 5,0 "></polygon>\n	</marker>\n\n	<marker id="mim-binding-end-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="-2" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="-12,6 0,0 -12,-6 -5,0 "></polygon>\n	</marker>\n\n	<!-- mim-modification markers: four-point polygon, no stroke -->\n\n	<marker id="mim-modification-start-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="0" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="12,6 0,0 12,-6 5,0 "></polygon>\n	</marker>\n\n	<marker id="mim-modification-end-black" fill="black" markerUnits="strokeWidth" markerWidth="12" markerHeight="12" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="-2" y="-0.6" width="2" height="1.2"></rect>\n	<polygon stroke-width="0" points="-12,6 0,0 -12,-6 -5,0 "></polygon>\n	</marker>\n\n	<!-- mim-catalysis markers: circle, drawing-board fill and black stroke -->\n\n	<marker id="mim-catalysis-start-black" class="drawing-board-color-fill" stroke="black" markerHeight="12" markerWidth="12" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<circle cx="5.3" cy="0" r="5.3px" stroke-width="1px"></circle>\n	</marker>\n\n	<marker id="mim-catalysis-end-black" class="drawing-board-color-fill" stroke="black" markerHeight="12" markerUnits="strokeWidth" markerWidth="12" orient="auto" refX="5" refY="0" viewBox="-6.5 -6 12 12">\n	<circle cx="-0.3" cy="0" r="5.3px" stroke-width="1px"></circle>\n	</marker>\n\n	<!-- mim-cleavage markers: two lines and extended drawing-board rect -->\n\n	<marker id="mim-cleavage-start-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="-8 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="3.5" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="3.7" y1="0" x2="3.7" y2="6"></line>	\n	<line fill="none" stroke-width=".4" x1="3.7" y1="6" x2="-8" y2="-6"></line>	\n	</marker>\n\n	<marker id="mim-cleavage-end-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="-4 -6 12 12">\n	<rect stroke="none" x="-3.5" y="-0.6" width="3.5" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="-3.7" y1="0" x2="-3.7" y2="-6"></line>	\n	<line fill="none" stroke-width=".4" x1="-3.7" y1="-6" x2="8" y2="6"></line>	\n	</marker>\n\n	<!-- mim-branching-left markers: line and extended drawing-board rect -->\n\n	<marker id="mim-branching-left-start-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="3.5" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="3.7" y1="0" x2="0" y2="-6"></line>	\n	</marker>\n\n	<marker id="mim-branching-left-end-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-3.5" y="-0.6" width="3.5" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="-3.7" y1="0" x2="0" y2="6"></line>	\n	</marker>\n\n	<!-- mim-branching-right markers: line and extended drawing-board rect -->\n\n	<marker id="mim-branching-right-start-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="3.5" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="3.7" y1="0" x2="0" y2="6"></line>\n	</marker>\n\n	<marker id="mim-branching-right-end-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-3.5" y="-0.6" width="3.5" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="-3.7" y1="0" x2="0" y2="-6"></line>	\n	</marker>\n\n	<!-- mim-transcription-translation markers: two lines and an open trigular polygon, plus extended drawing-board rect -->\n\n	<marker id="mim-transcription-translation-start-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="0" y="-0.6" width="6" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="9" y1="0" x2="9" y2="-4"></line>\n	<line fill="none" stroke-width=".4" x1="9" y1="-4" x2="5" y2="-4"></line>\n	<polygon stroke-width=".4" points="5,-6 0,-4 5,-2"></polygon>\n	</marker>\n\n	<marker id="mim-transcription-translation-end-black" class="drawing-board-color-fill" stroke="black" markerHeight="24" markerWidth="24" markerUnits="strokeWidth" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-6" y="-0.6" width="6" height="1.2"></rect>\n	<line fill="none" stroke-width=".4" x1="-9" y1="0" x2="-9" y2="4"></line>\n	<line fill="none" stroke-width=".4" x1="-9" y1="4" x2="-5" y2="4"></line>\n	<polygon stroke-width=".4" points="-5,6 0,4 -5,2"></polygon>	\n	</marker>\n	\n	<!-- mim-covalent-bond markers: not much to see here! -->\n\n	<marker id="mim-covalent-bond-start-black" markerUnits="strokeWidth" markerWidth="10" markerHeight="10" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	</marker>\n\n	<marker id="mim-covalent-bond-end-black" markerUnits="strokeWidth" markerWidth="10" markerHeight="10" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	</marker>\n\n\n	<!-- double-line-hack markers are used in double line handling; they include their own \n	special blank rect to obscure the ends -->\n\n	<marker id="double-line-hack-start" markerUnits="strokeWidth" markerWidth="10" markerHeight="10" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="0" y="-1.5" width="2.3" height="3"></rect>\n	</marker>\n\n	<marker id="double-line-hack-end" markerUnits="strokeWidth" markerWidth="10" markerHeight="10" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect class="drawing-board-color-fill" stroke="none" x="-2.3" y="-1.5" width="2.3" height="3"></rect>\n	</marker>\n\n\n	<!-- mim-gap markers: just an extended drawing-board rect -->\n	<!-- \n	TODO This could be refactored to make the shape match the viewbox.\n	It can overlap the side of the shape, blanking out a small part of it when the edge is at an angle.\n	-->\n\n	<marker id="mim-gap-start-black" class="drawing-board-color-fill" markerUnits="strokeWidth" markerWidth="10" markerHeight="10" orient="auto" refX="0" refY="0" viewBox="0 -6 12 12">\n	<rect stroke="none" x="-2" y="-0.7" width="8" height="1.4"></rect>\n	</marker>\n\n	<marker id="mim-gap-end-black" class="drawing-board-color-fill" markerUnits="strokeWidth" markerWidth="10" markerHeight="10" orient="auto" refX="0" refY="0" viewBox="-12 -6 12 12">\n	<rect stroke="none" x="-6" y="-0.7" width="8" height="1.4"></rect>\n	</marker>\n\n	<!-- ***************************\n	Symbols (shapes) \n	*************************** -->\n\n	<symbol id="none" class="shape-none" viewBox="0 0 100 50" preserveAspectRatio="none">\n	<desc>none</desc>\n	</symbol>\n\n	<!-- arc -->\n\n	<symbol id="arc" viewBox="0 0 100 100" preserveAspectRatio="none">\n	<desc>arc</desc>\n	<path d="m1.5,50.5c0,-16.16667 8.16667,-24.25 24.5,-24.25s24.5,-8.08334 24.5,-24.25c0,16.16666 8.16666,24.25 24.49999,24.25s24.50001,8.08333 24.50001,24.25" vector-effect="non-scaling-stroke"></path>	\n	</symbol> \n\n	<!-- brace -->\n\n	<clipPath id="brace-clip-path">\n	<path d="m1.5,49.499996c0,-16.166668 8.166666,-24.249996 24.499998,-24.249996s24.499998,-8.083336 24.499998,-24.250002c0,16.166666 8.166664,24.250002 24.499996,24.250002s24.5,8.083328 24.5,24.249996" vector-effect="non-scaling-stroke"></path>\n        </clipPath>\n\n	<symbol id="brace" viewBox="0 0 100 50" preserveAspectRatio="none">\n	<desc>brace</desc>\n	<path d="m1.5,49.499996c0,-16.166668 8.166666,-24.249996 24.499998,-24.249996s24.499998,-8.083336 24.499998,-24.250002c0,16.166666 8.166664,24.250002 24.499996,24.250002s24.5,8.083328 24.5,24.249996" style="clip-path: url(#brace-clip-path); " vector-effect="non-scaling-stroke"></path>\n  	</symbol>\n	\n	<!-- group of style "none" -->\n	<!-- see group.js for symbol definition. Note that the style definition is still in this SVG template file. -->\n\n	<!-- group of style "group" -->\n	<!-- see group.js for symbol definition. Note that the style definition is still in this SVG template file. -->\n\n	<!-- group of style "complex" -->\n	<!-- see group.js for symbol definition. Note that the style definition is still in this SVG template file. -->\n\n	<!-- group of style "pathway" -->\n	<!-- see group.js for symbol definition. Note that the style definition is still in this SVG template file. -->\n\n	<!-- Endoplasmic reticulum -->\n	<clipPath id="endoplasmic-reticulum-clip-path">\n	<path d="m73.52756,56.60967c-5.62457,-18.60675 23.51463,-32.43358 23.40173,-45.06604c-0.34426,-4.86102 -10.48934,-8.89743 -18.28974,-5.33395c-17.04119,7.87556 -15.64949,29.30503 -21.20533,42.23387c-0.35661,3.60951 -7.36274,2.46926 -7.74964,-0.48694c-5.8512,-11.38871 17.13534,-24.48692 5.96075,-29.42586c-19.63467,-8.16979 -28.75184,21.15346 -22.0682,28.81784c7.4956,14.17602 -2.17949,24.40679 -6.74689,15.49637c-2.44209,-5.30613 6.06605,-11.08445 -0.80351,-16.17689c-4.31991,-2.79993 -11.75555,-0.64618 -16.15468,3.0943c-12.89117,10.73799 4.72957,40.98145 20.96467,36.14635c4.69833,-1.95989 -3.23603,-8.70151 3.90717,-9.59951c7.29767,-0.81255 5.17628,6.18889 7.68745,9.22691c2.3071,4.0509 4.83232,8.35538 10.7626,11.6237c4.78642,2.53724 15.29437,2.11225 16.77148,-1.95795c2.0318,-9.26291 -26.11129,-28.35848 -10.68903,-31.2815c18.55524,-2.71473 4.74866,23.84573 24.31006,29.69419c9.50188,2.02824 15.63902,-0.62194 14.81255,-4.03272c-2.74586,-11.26327 -25.13557,-22.6802 -24.96441,-33.14968" vector-effect="non-scaling-stroke"></path>\n	</clipPath>\n	\n	<symbol id="endoplasmic-reticulum" viewBox="0 0 100 100" preserveAspectRatio="none">\n	<desc>endoplasmic-reticulum</desc>\n	<path d="m73.52756,56.60967c-5.62457,-18.60675 23.51463,-32.43358 23.40173,-45.06604c-0.34426,-4.86102 -10.48934,-8.89743 -18.28974,-5.33395c-17.04119,7.87556 -15.64949,29.30503 -21.20533,42.23387c-0.35661,3.60951 -7.36274,2.46926 -7.74964,-0.48694c-5.8512,-11.38871 17.13534,-24.48692 5.96075,-29.42586c-19.63467,-8.16979 -28.75184,21.15346 -22.0682,28.81784c7.4956,14.17602 -2.17949,24.40679 -6.74689,15.49637c-2.44209,-5.30613 6.06605,-11.08445 -0.80351,-16.17689c-4.31991,-2.79993 -11.75555,-0.64618 -16.15468,3.0943c-12.89117,10.73799 4.72957,40.98145 20.96467,36.14635c4.69833,-1.95989 -3.23603,-8.70151 3.90717,-9.59951c7.29767,-0.81255 5.17628,6.18889 7.68745,9.22691c2.3071,4.0509 4.83232,8.35538 10.7626,11.6237c4.78642,2.53724 15.29437,2.11225 16.77148,-1.95795c2.0318,-9.26291 -26.11129,-28.35848 -10.68903,-31.2815c18.55524,-2.71473 4.74866,23.84573 24.31006,29.69419c9.50188,2.02824 15.63902,-0.62194 14.81255,-4.03272c-2.74586,-11.26327 -25.13557,-22.6802 -24.96441,-33.14968" style="clip-path: url(#endoplasmic-reticulum-clip-path); " vector-effect="non-scaling-stroke"></path>\n	</symbol>\n\n	<!-- Golgi apparatus -->\n	<clipPath id="golgi-apparatus-clip-path1">\n	<path d="m58.46714,27.713327c-22.205345,-29.90079 37.310066,-30.258356 25.567245,-4.823446c-8.807655,18.581238 -17.066429,58.135235 -0.941673,99.22044c13.31469,27.066696 -41.748463,27.760925 -27.755554,-1.469849c11.345825,-29.420242 10.286858,-80.336422 3.129982,-92.927145z" vector-effect="non-scaling-stroke"></path>\n   	</clipPath> \n	\n	<clipPath id="golgi-apparatus-clip-path2">\n   	<path d="m31.214371,36.214363c-10.791712,-21.427903 29.897598,-19.848164 18.407501,0.670895c-4.066933,7.422386 -5.782803,61.572803 1.160713,75.028805c8.52943,18.597427 -32.852985,19.355408 -20.500162,-2.250633c6.952761,-17.358604 10.473742,-52.291187 0.931948,-73.449066z" vector-effect="non-scaling-stroke"></path>\n	</clipPath> \n	\n	<clipPath id="golgi-apparatus-clip-path3">\n   	<path d="m29.803959,52.160912c1.584177,11.474716 2.723461,16.737267 -1.482977,38.361366c-3.731956,12.989006 -3.600399,16.340691 -11.732334,19.412781c-6.683298,1.658531 -11.864832,-9.789436 -4.793299,-16.11377c4.855728,-5.623222 6.141087,-10.882362 6.658888,-22.954659c-0.239212,-9.521427 0.814508,-15.823826 -5.36692,-19.958626c-7.624315,-2.195171 -6.088041,-16.534611 4.824059,-13.863804c5.849354,1.027065 10.282408,8.561516 11.892582,15.116711z" vector-effect="non-scaling-stroke"></path>\n	</clipPath> \n	\n	<symbol id="golgi-apparatus" viewBox="0 0 90 150" preserveAspectRatio="none">\n	<desc>golgi-apparatus</desc>\n	<path d="m58.46714,27.713327c-22.205345,-29.90079 37.310066,-30.258356 25.567245,-4.823446c-8.807655,18.581238 -17.066429,58.135235 -0.941673,99.22044c13.31469,27.066696 -41.748463,27.760925 -27.755554,-1.469849c11.345825,-29.420242 10.286858,-80.336422 3.129982,-92.927145z" style="clip-path: url(#golgi-apparatus-clip-path1); " vector-effect="non-scaling-stroke"></path>\n   	<path d="m31.214371,36.214363c-10.791712,-21.427903 29.897598,-19.848164 18.407501,0.670895c-4.066933,7.422386 -5.782803,61.572803 1.160713,75.028805c8.52943,18.597427 -32.852985,19.355408 -20.500162,-2.250633c6.952761,-17.358604 10.473742,-52.291187 0.931948,-73.449066z" style="clip-path: url(#golgi-apparatus-clip-path2); " vector-effect="non-scaling-stroke"></path>\n   	<path d="m29.803959,52.160912c1.584177,11.474716 2.723461,16.737267 -1.482977,38.361366c-3.731956,12.989006 -3.600399,16.340691 -11.732334,19.412781c-6.683298,1.658531 -11.864832,-9.789436 -4.793299,-16.11377c4.855728,-5.623222 6.141087,-10.882362 6.658888,-22.954659c-0.239212,-9.521427 0.814508,-15.823826 -5.36692,-19.958626c-7.624315,-2.195171 -6.088041,-16.534611 4.824059,-13.863804c5.849354,1.027065 10.282408,8.561516 11.892582,15.116711z" style="clip-path: url(#golgi-apparatus-clip-path3); " vector-effect="non-scaling-stroke"></path>\n	</symbol>\n\n	<!-- this is the dragbox shape for any set of elements that have been grouped. this only shows up in edit mode. -->\n\n	<symbol id="group-drag-box" viewBox="0 0 120 180" preserveAspectRatio="none">\n	<desc>Group</desc>\n	<rect x="0" y="0" width="120" height="180" vector-effect="non-scaling-stroke"></rect>\n	</symbol>\n\n	<!-- Shape is placeholder until it can be replaced with correct shape. -->\n	<clipPath id="hexagon-clip-path">\n	<path d="m1.42004,50.99635l21.07262,-42.13943l56.19152,0l21.0667,42.13943l-21.0667,42.14507l-56.19152,0l-21.07262,-42.14507z" vector-effect="non-scaling-stroke"></path>\n        </clipPath>\n      \n	<symbol id="hexagon" viewBox="0 0 100 100" preserveAspectRatio="none">\n	<desc>hexagon</desc>\n	<path d="m1.42004,50.99635l21.07262,-42.13943l56.19152,0l21.0667,42.13943l-21.0667,42.14507l-56.19152,0l-21.07262,-42.14507z" style="clip-path: url(#hexagon-clip-path); " vector-effect="non-scaling-stroke"></path>\n	</symbol>\n\n	<!-- Shape is placeholder until it can be replaced with correct shape. -->\n	<!-- The line part of this shape has a stroke width set to 1, to make it the same as the circle after clip path is applied -->\n	<clipPath id="mim-degradation-circle-clip-path">\n	<circle cx="50" cy="50" r="49" vector-effect="non-scaling-stroke"></circle>\n  	</clipPath>\n  	\n	<symbol id="mim-degradation" viewBox="0 0 100 100" preserveAspectRatio="none">\n	<desc>mim-degradation</desc>\n	<circle cx="50" cy="50" r="49" style="clip-path: url(#mim-degradation-circle-clip-path); " vector-effect="non-scaling-stroke"></circle>\n  	<line x1="1" y1="1" x2="100" y2="100" stroke-width="1" vector-effect="non-scaling-stroke"></line>\n  	</symbol>\n	\n	<!-- mitochondria -->\n	<clipPath id="mitochondria-clip-path">\n	<path d="m14.894899,26.347357c4.363817,-0.741571 3.827518,17.036169 8.182638,16.183825c8.27347,0.030762 2.982006,-28.148991 9.899754,-28.336687c6.967995,-0.187704 2.246651,29.947527 9.204983,29.43981c7.632813,-0.560024 0.507309,-32.935357 8.136253,-33.623082c7.698521,-0.689259 2.919197,32.039941 10.628349,32.224557c6.546684,0.160011 3.026451,-27.642808 9.56057,-26.921232c7.192177,0.79388 0.664818,29.842905 7.781624,31.667604c4.748405,1.215439 4.420822,-18.257757 9.204018,-17.440804c11.128883,7.577278 8.628105,37.698658 -2.179977,44.645138c-3.138542,0.698479 -3.965698,-10.502029 -7.112938,-9.905075c-5.59005,1.058502 -3.982124,22.284088 -9.603096,21.799461c-5.239281,-0.456947 -2.226364,-21.636383 -7.47047,-21.730232c-6.961235,-0.116928 -3.357895,28.924408 -10.316231,28.495148c-6.140846,-0.375397 -1.73064,-24.950363 -7.825104,-26.191963c-5.681847,-1.156982 -5.378429,22.170242 -11.027426,20.680939c-6.249069,-1.644684 -0.469624,-26.673519 -6.759275,-27.865887c-3.728954,-0.706188 -2.647665,14.400654 -6.403677,14.545292c-14.016198,-5.938736 -15.748776,-39.707981 -3.899994,-47.666811z" vector-effect="non-scaling-stroke"></path>\n	</clipPath> \n	\n	<symbol id="mitochondria" viewBox="0 0 100 100" preserveAspectRatio="none">\n	<desc>Mitochondria</desc>\n	<use x="0" y="0" width="100" height="100" xlink:href="#oval"></use>\n	<path d="m14.894899,26.347357c4.363817,-0.741571 3.827518,17.036169 8.182638,16.183825c8.27347,0.030762 2.982006,-28.148991 9.899754,-28.336687c6.967995,-0.187704 2.246651,29.947527 9.204983,29.43981c7.632813,-0.560024 0.507309,-32.935357 8.136253,-33.623082c7.698521,-0.689259 2.919197,32.039941 10.628349,32.224557c6.546684,0.160011 3.026451,-27.642808 9.56057,-26.921232c7.192177,0.79388 0.664818,29.842905 7.781624,31.667604c4.748405,1.215439 4.420822,-18.257757 9.204018,-17.440804c11.128883,7.577278 8.628105,37.698658 -2.179977,44.645138c-3.138542,0.698479 -3.965698,-10.502029 -7.112938,-9.905075c-5.59005,1.058502 -3.982124,22.284088 -9.603096,21.799461c-5.239281,-0.456947 -2.226364,-21.636383 -7.47047,-21.730232c-6.961235,-0.116928 -3.357895,28.924408 -10.316231,28.495148c-6.140846,-0.375397 -1.73064,-24.950363 -7.825104,-26.191963c-5.681847,-1.156982 -5.378429,22.170242 -11.027426,20.680939c-6.249069,-1.644684 -0.469624,-26.673519 -6.759275,-27.865887c-3.728954,-0.706188 -2.647665,14.400654 -6.403677,14.545292c-14.016198,-5.938736 -15.748776,-39.707981 -3.899994,-47.666811z" style="clip-path: url(#mitochondria-clip-path); " vector-effect="non-scaling-stroke"></path>\n	</symbol>\n\n	<!-- oval -->\n        <clipPath id="oval-clip-path">\n      	<ellipse cx="50" cy="25" rx="50" ry="25" vector-effect="non-scaling-stroke"></ellipse>\n        </clipPath>\n\n	<symbol id="oval" viewBox="0 0 100 50" preserveAspectRatio="none">\n	<desc>Ellipse</desc>\n	<ellipse cx="50" cy="25" rx="50" ry="25" style="clip-path: url(#oval-clip-path); " vector-effect="non-scaling-stroke"></ellipse>\n	</symbol>\n\n	<!-- pentagon -->\n        <clipPath id="pentagon-clip-path">\n	<polygon points="59.159732818603516,99.61322021484375 95,50.28331756591797 59.159732818603516,0.9534196853637695 1.168962001800537,19.795764923095703 1.168962001800537,80.7708740234375 " vector-effect="non-scaling-stroke"></polygon>\n	</clipPath>\n	\n	<symbol id="pentagon" viewBox="0 0 100 100" preserveAspectRatio="none">\n	<desc>pentagon</desc>\n	<polygon points="59.159732818603516,99.61322021484375 95,50.28331756591797 59.159732818603516,0.9534196853637695 1.168962001800537,19.795764923095703 1.168962001800537,80.7708740234375 " style="clip-path: url(#pentagon-clip-path); " vector-effect="non-scaling-stroke"></polygon>\n	</symbol>\n\n	<!-- rectangle -->\n      	<clipPath id="rectangle-clip-path">\n	<rect x="0" y="0" width="100" height="50" vector-effect="non-scaling-stroke"></rect>\n      	</clipPath>\n      \n	<symbol id="rectangle" viewBox="0 0 100 50" preserveAspectRatio="none">\n	<desc>Rectangle</desc>\n	<rect x="0" y="0" width="100" height="50" style="clip-path: url(#rectangle-clip-path); " vector-effect="non-scaling-stroke"></rect>\n	</symbol>\n\n	<!-- rounded-rectangle -->\n	<!-- TODO the rounded corners need to be constant in size, meaning they will probably have to be defined in JavaScript. -->\n      	<clipPath id="rounded-rectangle-clip-path">\n	<rect x="0" y="0" rx="2.5" ry="2.5" width="100" height="50" vector-effect="non-scaling-stroke"></rect>\n      	</clipPath>\n      \n	<symbol id="rounded-rectangle" viewBox="0 0 100 50" preserveAspectRatio="none">\n	<desc>Rounded Rectangle</desc>\n	<rect x="0" y="0" rx="2.5" ry="2.5" width="100" height="50" overflow="hidden" vector-effect="non-scaling-stroke"></rect>\n	</symbol>\n\n	<!-- Sarcoplasmic reticulum -->\n	<clipPath id="sarcoplasmic-reticulum-clip-path">\n	<path d="m46.60182,1.40724c-32.37224,1.34138 -36.32004,22.77011 -26.50318,38.12777c9.31826,18.3425 -18.7656,30.15016 2.56955,49.37807c16.82126,13.11594 46.33175,6.10508 52.12638,-8.56826c5.89916,-15.24847 -10.95099,-26.0272 -3.29316,-40.96135c10.85342,-19.88432 -0.77615,-38.13043 -24.89959,-37.97624z" vector-effect="non-scaling-stroke"></path>	\n	</clipPath>\n	\n	<symbol id="sarcoplasmic-reticulum" viewBox="0 0 80 100" preserveAspectRatio="none">\n	<desc>sarcoplasmic-reticulum</desc>\n	<path d="m46.60182,1.40724c-32.37224,1.34138 -36.32004,22.77011 -26.50318,38.12777c9.31826,18.3425 -18.7656,30.15016 2.56955,49.37807c16.82126,13.11594 46.33175,6.10508 52.12638,-8.56826c5.89916,-15.24847 -10.95099,-26.0272 -3.29316,-40.96135c10.85342,-19.88432 -0.77615,-38.13043 -24.89959,-37.97624z" style="clip-path: url(#sarcoplasmic-reticulum-clip-path); " vector-effect="non-scaling-stroke"></path>	\n	</symbol>\n\n	<!-- triangle -->\n\n	<clipPath id="triangle-clip-path">\n	<polygon points="1,49 49,24 1,1" vector-effect="non-scaling-stroke"></polygon>\n        </clipPath>\n\n	<symbol id="triangle" viewBox="0 0 50 50" preserveAspectRatio="none">\n	<desc>triangle</desc>\n	<polygon points="1,49 49,24 1,1" style="clip-path: url(#triangle-clip-path);" vector-effect="non-scaling-stroke"></polygon>\n	</symbol>\n\n	<!-- TODO Update placeholders for SVG definitions of shapes.\n\n	For current symbols in GIF and/or BMP format, see\n	http://svn.bigcat.unimaas.nl/pathvisio/trunk/modules/org.pathvisio.core/resources/\n\n	Should we add SBGN? Available as SVG here: http://www.sbgn.org/Documents/Templates \n	Not sure about adding lines. Should they be only defined in CSS?\n	-->\n\n	</defs>\n		<g id="viewport">\n    </g>\n\n	</svg>\n</div>\n    <div id="viewertoolbar" style="position:absolute; right: 0px; top: 20px;">\n      <div class="ui-widget ui-corner-all" style="position: absolute; right: 5px; top: 5px; z-index: 1001; background-color: rgb(221, 221, 221); border: 1px solid rgb(170, 170, 170); width:130px">\n        <span class="icon icon-eye-open"></span>\n        <span class="twitter-typeahead" style="position: relative; display: inline-block;"><input class="tt-hint" type="text" autocomplete="off" spellcheck="off" disabled="" style="position: absolute; top: 0px; left: 0px; border-color: transparent; box-shadow: none; background-attachment: scroll; background-clip: border-box; background-color: rgb(255, 255, 255); background-image: none; background-origin: padding-box; background-size: auto; background-position: 0% 0%; background-repeat: repeat repeat;"><input id="highlight-by-label" placeholder="Find in pathway" class="ui-autocomplete-input ui-corner-all tt-query" autocomplete="off" role="textbox" aria-autocomplete="list" aria-haspopup="true" style="float: right; width: 100px; position: relative; vertical-align: top; background-color: transparent;" spellcheck="false" dir="auto"><span style="position: absolute; left: -9999px; visibility: hidden; white-space: nowrap; font-family: \'Lucida Grande\'; font-size: 11px; font-style: normal; font-variant: normal; font-weight: 400; word-spacing: 0px; letter-spacing: 0px; text-indent: 0px; text-rendering: auto; text-transform: none;"></span><span class="tt-dropdown-menu" style="position: absolute; top: 100%; left: 0px; z-index: 100; display: none;"></span></span>\n        <div class="ui-corner-all" style="position: absolute; font-size: 75%; background-color: white; display: none;">\n        </div>\n      </div>\n      <!-- see http://bumbu.github.io/cytoscape.js/debug/ for example of cytoscape.js -->\n      <div class="ui-cytoscape-panzoom">\n          <div class="ui-cytoscape-panzoom-zoom-in ui-cytoscape-panzoom-zoom-button">\n            <span class="icon icon-plus"></span>\n          </div>\n          <div class="ui-cytoscape-panzoom-zoom-out ui-cytoscape-panzoom-zoom-button">\n            <span class="icon icon-minus"></span>\n          </div>\n          <div class="ui-cytoscape-panzoom-reset ui-cytoscape-panzoom-zoom-button">\n            <span class="icon icon-resize-full"></span>\n          </div>\n          <div class="ui-cytoscape-panzoom-slider">\n              <div class="ui-cytoscape-panzoom-slider-background">\n              </div>\n              <div class="ui-cytoscape-panzoom-slider-handle" style="top: 42.80000001192093px;">\n                <span class="icon icon-minus"></span>\n              </div>\n              <div class="ui-cytoscape-panzoom-no-zoom-tick" style="top: 42.80000001192093px;">\n              </div>\n          </div>\n          <div class="ui-cytoscape-panzoom-panner">\n              <div class="ui-cytoscape-panzoom-panner-handle">\n              </div>\n              <div class="ui-cytoscape-panzoom-pan-up ui-cytoscape-panzoom-pan-button">\n              </div>\n              <div class="ui-cytoscape-panzoom-pan-down ui-cytoscape-panzoom-pan-button">\n              </div>\n              <div class="ui-cytoscape-panzoom-pan-left ui-cytoscape-panzoom-pan-button">\n              </div>\n              <div class="ui-cytoscape-panzoom-pan-right ui-cytoscape-panzoom-pan-button">\n              </div>\n              <div class="ui-cytoscape-panzoom-pan-indicator" style="display: none; left: 22.424611085682006px; top: 0.12287108520014556px; background-color: rgb(127, 127, 127); background-position: initial initial; background-repeat: initial initial;">\n              </div>\n          </div>\n      </div>\n    </div>\n    <div id="details-frame" style="visibility: hidden; position:absolute; right: 75px; top: 100px;" class="data-node ui-draggable">\n    </div>\n  </div>\n</div>\n',pathvisio.helpers=function(){function isUrl(str){var urlPattern=new RegExp("((http|https)(://))?([a-zA-Z0-9]+[.]{1}){2}[a-zA-z0-9]+(/{1}[a-zA-Z0-9]+)*/?","i");
return urlPattern.test(str)}function splitStringByNewLine(str){return str.split(/\r\n|\r|\n/g)}function cloneNode(selector){var node=d3.select(selector).node();return d3.select(node.parentNode.insertBefore(node.cloneNode(!0),node.nextSibling))}function getUrlParam(name){var parameter=decodeURIComponent((new RegExp("[?|&]"+name+"="+"([^&;]+?)(&|#|;|$)").exec(location.search)||[,""])[1].replace(/\+/g,"%20"))||null;return parameter?parameter:(console.warn('Warning: URL parameter "'+name+'" is null.'),null)}function convertToArray(object){var array=null;return"[object Object]"===Object.prototype.toString.call(object)?(array=[],array.push(object),array):"[object Array]"===Object.prototype.toString.call(object)?object:"[object String]"===Object.prototype.toString.call(object)?(array=[],array.push(object),array):void 0}function getWindowDimensions(){var winW=630,winH=460;return document.body&&document.body.offsetWidth&&(winW=document.body.offsetWidth,winH=document.body.offsetHeight),"CSS1Compat"==document.compatMode&&document.documentElement&&document.documentElement.offsetWidth&&(winW=document.documentElement.offsetWidth,winH=document.documentElement.offsetHeight),window.innerWidth&&window.innerHeight&&(winW=window.innerWidth,winH=window.innerHeight),{width:winW,height:winH}}function moveArrayItem(arr,old_index,new_index){if(new_index>=arr.length)for(var k=new_index-arr.length;k--+1;)arr.push(void 0);return arr.splice(new_index,0,arr.splice(old_index,1)[0]),arr}return{isUrl:isUrl,splitStringByNewLine:splitStringByNewLine,getUrlParam:getUrlParam,cloneNode:cloneNode,convertToArray:convertToArray,getWindowDimensions:getWindowDimensions,moveArrayItem:moveArrayItem}}(),pathvisio.pathway=function(){function gpml2json(gpml,callback){console.log("GPML"),console.log(gpml),pathway=xml.xmlToJSON(gpml,!0).pathway,console.log("raw json from xml2json"),console.log(xml.xmlToJSON(gpml,!0).pathway);var xmlns=null;try{xmlns=pathway.xmlns}catch(e){return console.log(e.message),void 0}var gpmlXmlnsSupported="http://pathvisio.org/GPML/2013a",gpmlXmlnsIdentifier="/GPML/";if(-1===xmlns.indexOf(gpmlXmlnsIdentifier))return alert("Pathvisio.js does not support the data format provided. Please convert to GPML and retry."),console.log("Pathvisio.js does not support the data format provided. Please convert to GPML and retry."),void 0;xmlns!=gpmlXmlnsSupported&&alert("Pathvisio.js may not fully support the version of GPML provided (xmlns: "+xmlns+"). Please convert to the supported version of GPML (xmlns: "+gpmlXmlnsSupported+")."),pathway.boardWidth=pathway.graphics.boardWidth,pathway.boardHeight=pathway.graphics.boardHeight,pathway.infoBox.x=0,delete pathway.infoBox.centerX,pathway.infoBox.y=0,delete pathway.infoBox.centerY;try{pathway.hasOwnProperty("comment")?(pathway.comments=pathvisio.helpers.convertToArray(pathway.comment),delete pathway.comment,pathway.comments.forEach(function(){})):console.log("No element(s) named 'comment' found in this gpml file.")}catch(e){console.log("Error converting comment to json: "+e.message)}try{pathway.hasOwnProperty("group")?(pathway.groups=pathvisio.helpers.convertToArray(pathway.group),delete pathway.group,pathway.groups.forEach(function(element){element.style=element.hasOwnProperty("style")?element.style.toLowerCase():"none"})):console.log("No element(s) named 'group' found in this gpml file.")}catch(e){console.log("Error converting group to json: "+e.message)}try{if(pathway.hasOwnProperty("graphicalLine")){var graphicalLines=pathvisio.helpers.convertToArray(pathway.graphicalLine);delete pathway.graphicalLine,void 0===pathway.edges&&(pathway.edges=[]),graphicalLines.forEach(function(element){element.edgeType="graphical-line",pathway.edges.push(element)})}else console.log("No element(s) named 'graphicalLine' found in this gpml file.")}catch(e){console.log("Error converting graphicalLine to json: "+e.message)}try{if(pathway.hasOwnProperty("interaction")){var interactions=pathvisio.helpers.convertToArray(pathway.interaction);delete pathway.interaction,void 0===pathway.edges&&(pathway.edges=[]),interactions.forEach(function(element){element.edgeType="interaction",pathway.edges.push(element)}),pathway.edges}else console.log("No element(s) named 'interaction' found in this gpml file.")}catch(e){console.log("Error converting interaction to json: "+e.message)}try{pathway.hasOwnProperty("edges")?pathway.edges=pathvisio.pathway.edge.gpml2json(pathway.edges):console.log("No element(s) named 'edges' found in this gpml file.")}catch(e){console.log("Error converting edges to json: "+e.message)}try{if(pathway.hasOwnProperty("dataNode")){var dataNodes=pathvisio.helpers.convertToArray(pathway.dataNode);delete pathway.dataNode,dataNodes.forEach(function(element){element.elementType="data-node",element.dataNodeType=caseConverter.paramCase(element.type),delete element.type,element.hasOwnProperty("xref")&&(element.xref.database||element.xref.iD?(element.xRef=element.xref,delete element.xref,element.xRef.id=element.xRef.iD,delete element.xRef.iD):delete element.xref)}),pathway.nodes=pathway.hasOwnProperty("nodes")?pathway.nodes.concat(dataNodes):dataNodes}else console.log("No element(s) named 'dataNode' found in this gpml file.")}catch(e){console.log("Error converting dataNode to json: "+e.message)}try{if(pathway.hasOwnProperty("label")){var labels=pathvisio.helpers.convertToArray(pathway.label);delete pathway.label,labels.forEach(function(element){element.elementType="label"}),pathway.nodes=pathway.hasOwnProperty("nodes")?pathway.nodes.concat(labels):labels}else console.log("No element(s) named 'label' found in this gpml file.")}catch(e){console.log("Error converting label to json: "+e.message)}try{if(pathway.hasOwnProperty("shape")){var shapes=pathvisio.helpers.convertToArray(pathway.shape);delete pathway.shape,shapes.forEach(function(element){element.elementType="shape"}),pathway.nodes=pathway.hasOwnProperty("nodes")?pathway.nodes.concat(shapes):shapes}else console.log("No element(s) named 'shape' found in this gpml file.")}catch(e){console.log("Error converting shape to json: "+e.message)}try{pathway.hasOwnProperty("nodes")?pathway.nodes=pathvisio.pathway.node.gpml2json(pathway.nodes):console.log("No element(s) named 'nodes' found in this gpml file.")}catch(e){console.log("Error converting nodes to json: "+e.message)}try{pathway.hasOwnProperty("biopaxRef")?(pathway.biopaxRefs=pathvisio.helpers.convertToArray(pathway.biopaxRef),delete pathway.biopaxRef):console.log("No element(s) named 'biopaxRef' for the element 'pathway' found in this gpml file.")}catch(e){console.log("Error converting biopaxRef to json: "+e.message)}try{pathway.hasOwnProperty("biopax")?(pathway.biopax.bpPublicationXrefs=pathvisio.helpers.convertToArray(pathway.biopax.bpPublicationXref),delete pathway.biopax.bpPublicationXref):console.log("No element(s) named 'biopax' found in this gpml file.")}catch(e){console.log("Error converting biopax to json: "+e.message)}console.log("JSON:"),console.log(pathway),console.log("pathway"),console.log(pathway),delete pathway.graphics,callback(pathway)}function getJson(url,callback){if(!url){var error="Error: URL not specified.";return console.warn(error),error}$.get(url,"application/xml",function(gpml){pathvisio.pathway.gpml2json(gpml,function(json){callback(json)})})}function highlightByLabel(nodeLabel){console.log("nodeLabel"),console.log(nodeLabel),svg.selectAll("g.nodes-container").attr("style","");var dataNodes=pathway.nodes.filter(function(element){return"data-node"===element.elementType}),dataNodesWithText=dataNodes.filter(function(element){return!!element.textLabel}),selectedNodes=dataNodesWithText.filter(function(element){return-1!==element.textLabel.text.indexOf(nodeLabel)});selectedNodes.forEach(function(node){console.log("node"),console.log(node);var nodeDomElement=svg.select("#nodes-container-"+node.graphId);nodeDomElement.attr("style","fill:yellow"),console.log("nodeDomElement"),console.log(nodeDomElement)})}function draw(svg,pathway){function dragmove(d){d.x=d3.event.x,d.y=d3.event.y,d3.select(this).attr("x",d3.event.x).attr("y",d3.event.y)}return pathway?(d3.behavior.drag().on("drag",dragmove),svg.attr("width",pathway.boardWidth),svg.attr("height",pathway.boardHeight),pathway.biopaxRefs&&svg.select("#viewport").selectAll(".pathway-publication-xref-text").data(pathway.biopaxRefs).enter().append("text").attr("id",function(d){return"pathway-publication-xref-text-"+d}).attr("x",0).attr("y",0).attr("transform",function(d,i){return"translate("+(200+12*i)+" "+12+")"}).attr("class","pathway-publication-xref-text").attr("style","").text(function(d){var index=0,rdfId=null;do rdfId=pathway.biopax.bpPublicationXrefs[index].rdfId,index+=1;while(rdfId!==d.Text&&index<pathway.biopax.bpPublicationXrefs.length);return index}),pathway.hasOwnProperty("groups")&&pathvisio.pathway.group.drawAll(svg,pathway),pathway.hasOwnProperty("edges")&&pathvisio.pathway.edge.drawAll(svg,pathway),pathway.hasOwnProperty("nodes")&&pathvisio.pathway.node.drawAll(svg,pathway),pathway.hasOwnProperty("infoBox")&&pathvisio.pathway.infoBox.draw(svg,pathway),void 0):(console.warn("Error: No data entered as input."),"Error")}function appendCustomShape(customShape,callback){img=document.createElement("img"),img.src=customShape.url,img.onload=function(){def=svg.select("defs").select("#"+customShape.id),def[0][0]?def.selectAll("*").remove():def=d3.select("svg").select("defs").append("symbol").attr("id",customShape.id).attr("viewBox","0 0 "+this.width+" "+this.height).attr("preserveAspectRatio","none"),dimensions=def.attr("viewBox").split(" "),def.append("image").attr("xlink:xlink:href",customShape.url).attr("x",dimensions[0]).attr("y",dimensions[1]).attr("width",dimensions[2]).attr("height",dimensions[3]),callback(null)}}function loadCustomShapes(args,callback){args.customShapes&&async.each(args.customShapes,appendCustomShape,function(){callback(null)})}function loadPartials(args,callback){async.series([function(callbackInside){console.log("2"),args.containerElement.html(pathvisioNS["tmp/pathvisio-js.html"]),svg=args.containerElement.select("#pathway-image"),callbackInside(null)},function(callbackInside){console.log("3"),loadCustomShapes(args,function(){callbackInside(null)})},function(callbackInside){console.log("4"),args.cssUrl?d3.text(args.cssUrl,"text/css",function(data){var defs=svg.select("defs"),style=defs.append("style").attr("type","text/css");style.text(data),callbackInside(null)}):callbackInside(null)}],function(err){console.log(err),callback()})}function load(args){return args.gpmlUrl?args.container?(args.containerElement=d3.select(args.container),1!==args.containerElement.length?console.warn("Error: Container selector must be matched by exactly one element."):(async.parallel([function(callback){console.log("1a"),loadPartials(args,callback)},function(callback){console.log("1b"),getJson(args.gpmlUrl,callback)}],function(err){console.log("5"),console.log(err),draw(svg,pathway);var nodeLabels=[];pathway.nodes.forEach(function(node){node.textLabel&&"data-node"===node.elementType&&nodeLabels.push(node.textLabel.text)}),$("#highlight-by-label").typeahead({name:"Find in pathway",local:nodeLabels,limit:10}),$(".icon-eye-open").click(function(){var nodeLabel=$("#highlight-by-label").val();nodeLabel?pathvisio.pathway.highlightByLabel(nodeLabel):console.warn("Error: No data node value entered.")})}),void 0)):console.warn("Error: No container selector specified as target for pathvisio.js."):console.warn("Error: No gpml URL specified as data source for pathvisio.js.")}var svg=null,pathway=null;return{draw:draw,load:load,getJson:getJson,gpml2json:gpml2json,highlightByLabel:highlightByLabel}}(),pathvisio.pathway.group=function(){function drawAll(svg,pathway){if(!svg||!pathway)return console.warn("Error: Missing input parameters.");var groups=pathway.groups;if(!groups)return console.warn("Error: No group data available.");var validGroups=groups.filter(function(el){var groupId=el.groupId;return pathway.nodes.filter(function(el){return el.groupRef===groupId}).length>0}),pathData=null;svg.select("#viewport").selectAll("use.group").data(validGroups).enter().append("path").attr("id",function(d){return"group-"+d.graphId}).attr("class",function(d){return"group group-"+d.style}).attr("d",function(d){var groupDimensions=getDimensions(pathway,d.groupId);return pathData="none"===d.style||"group"===d.style||"pathway"===d.style?"M "+groupDimensions.x+" "+groupDimensions.y+" L "+(groupDimensions.x+groupDimensions.width)+" "+groupDimensions.y+" L "+(groupDimensions.x+groupDimensions.width)+" "+(groupDimensions.y+groupDimensions.height)+" L "+groupDimensions.x+" "+(groupDimensions.y+groupDimensions.height)+" Z":"complex"===d.style?"M "+(groupDimensions.x+20)+" "+groupDimensions.y+" L "+(groupDimensions.x+groupDimensions.width-20)+" "+groupDimensions.y+" L "+(groupDimensions.x+groupDimensions.width)+" "+(groupDimensions.y+20)+" L "+(groupDimensions.x+groupDimensions.width)+" "+(groupDimensions.y+groupDimensions.height-20)+" L "+(groupDimensions.x+groupDimensions.width-20)+" "+(groupDimensions.y+groupDimensions.height)+" L "+(groupDimensions.x+20)+" "+(groupDimensions.y+groupDimensions.height)+" L "+groupDimensions.x+" "+(groupDimensions.y+groupDimensions.height-20)+" L "+groupDimensions.x+" "+(groupDimensions.y+20)+" Z":"M "+groupDimensions.x+" "+groupDimensions.y+" L "+(groupDimensions.x+groupDimensions.width)+" "+groupDimensions.y+" L "+(groupDimensions.x+groupDimensions.width)+" "+(groupDimensions.y+groupDimensions.height)+" L "+groupDimensions.x+" "+(groupDimensions.y+groupDimensions.height)+" Z"})}function getDimensions(pathway,groupId){var groupMembers=pathway.nodes.filter(function(el){return el.groupRef===groupId}),group={},margin=12;return group.x=d3.min(groupMembers,function(el){return el.x})-margin,group.y=d3.min(groupMembers,function(el){return el.y})-margin,group.width=d3.max(groupMembers,function(el){return el.x+el.width})-group.x+margin,group.height=d3.max(groupMembers,function(el){return el.y+el.height})-group.y+margin,group}return{drawAll:drawAll,getDimensions:getDimensions}}(),pathvisio.pathway.infoBox=function(){function draw(svg,pathway){if(!svg||!pathway)return console.warn("Error: Missing input parameters.");var infoBox=[];pathway.hasOwnProperty("name")&&infoBox.push({key:"Title",value:pathway.name}),pathway.hasOwnProperty("license")&&infoBox.push({key:"Availability",value:pathway.license}),pathway.hasOwnProperty("lastModified")&&infoBox.push({key:"Last modified",value:pathway.lastModified}),pathway.hasOwnProperty("organism")&&infoBox.push({key:"Organism",value:pathway.organism});var infoBoxElements=svg.select("#viewport").selectAll("text.info-box").data(infoBox).enter().append("text").attr("id",function(d,i){return"info-box-"+i}).attr("class","info-box").attr("x",0).attr("y",function(d,i){return 14+14*i});infoBoxElements.append("tspan").attr("class","info-box-property-name").text(function(d){return d.key+": "}),infoBoxElements.append("tspan").text(function(d){return d.value})}return{draw:draw}}(),pathvisio.pathway.node=function(){function gpml2json(rawJsonNodes){try{rawJsonNodes.forEach(function(element){if(element.hasOwnProperty("comment")&&(element.comments=pathvisio.helpers.convertToArray(element.comment),delete element.comment),element.graphics.hasOwnProperty("zorder")&&(element.zIndex=parseFloat(element.graphics.zorder)),element.x=parseFloat(element.graphics.centerX)-parseFloat(element.graphics.width)/2,element.y=parseFloat(element.graphics.centerY)-parseFloat(element.graphics.height)/2,element.width=parseFloat(element.graphics.width),element.height=parseFloat(element.graphics.height),element.graphics.hasOwnProperty("color")){var color=new RGBColor(element.graphics.color);color.ok?element.stroke=color.toHex():(console.warn("Invalid Color encountered. Setting Color to black."),element.fill="#000000")}if(element.symbolType=element.graphics.hasOwnProperty("shapeType")?caseConverter.paramCase(element.graphics.shapeType):"data-node"===element.elementType?"rectangle":"none",element.graphics.hasOwnProperty("fillColor"))if(element.graphics.fillColor=element.graphics.fillColor.toLowerCase(),"transparent"===element.graphics.fillColor)element.fillOpacity=0;else{var fill=new RGBColor(element.graphics.fillColor);fill.ok?element.fill=fill.toHex():(console.warn("Invalid FillColor encountered. Setting FillColor to gray."),element.fill="#999999"),"none"!==element.symbolType&&(element.fillOpacity=1)}if(element.graphics.hasOwnProperty("lineThickness")&&(element.strokeWidth=element.graphics.lineThickness),element.graphics.hasOwnProperty("lineStyle")&&(element.strokeStyle=element.graphics.lineStyle.toLowerCase(),"broken"===element.strokeStyle&&(element.strokeStyle="dashed")),element.hasOwnProperty("attribute")&&(element.attributes=pathvisio.helpers.convertToArray(element.attribute),delete element.attribute,element.attributes.forEach(function(el){"org.pathvisio.DoubleLineProperty"===el.key&&"Double"===el.value?(console.log("double"),console.log(el),element.strokeStyle="double"):"org.pathvisio.CellularComponentProperty"===el.key&&"None"!==el.value&&(element.cellularComponent=el.value)}),delete element.attributes),element.graphics.hasOwnProperty("rotation")&&(element.rotation=element.graphics.rotation*(180/Math.PI)),element.hasOwnProperty("textLabel"))if(element.textLabel){var text=element.textLabel.toString().replace("&#xA;","\r\n");delete element.textLabel,element.textLabel={},element.textLabel.text=text,element.hasOwnProperty("stroke")&&(element.textLabel.fill=element.stroke);var fontSize=null;fontSize=element.graphics.hasOwnProperty("fontSize")?element.graphics.fontSize:10,element.textLabel.fontSize=fontSize,element.graphics.hasOwnProperty("fontName")&&(element.textLabel.fontFamily=element.graphics.fontName),element.graphics.hasOwnProperty("fontWeight")&&(element.textLabel.fontWeight=element.graphics.fontWeight.toLowerCase()),element.graphics.hasOwnProperty("fontStyle")&&(element.textLabel.fontStyle=element.graphics.fontStyle.toLowerCase()),element.textLabel.textAnchor=alignToAnchorMappings.hasOwnProperty(element.graphics.align)?alignToAnchorMappings[element.graphics.align]:"middle",element.textLabel.vAlign=element.graphics.hasOwnProperty("valign")?element.graphics.valign.toLowerCase():"top"}else delete element.textLabel;try{element.hasOwnProperty("biopaxRef")?(element.biopaxRefs=pathvisio.helpers.convertToArray(element.biopaxRef),delete element.biopaxRef):console.log("No element(s) named 'biopaxRef' found for this node in this gpml file.")}catch(e){console.log("Error converting node's biopaxRef to json: "+e.message)}delete element.graphics});var validJsonNodes=rawJsonNodes.sort(function(a,b){return a.zIndex-b.zIndex});return validJsonNodes}catch(e){return console.log("Error converting labelable elements to json: "+e.message),e}}function drawAll(svg,pathway){if(!svg||!pathway)return svg||console.log("svg"),pathway||console.log("pathway"),console.warn("Error: Missing one or more required parameters: svg, pathway.");var nodesContainer=svg.select("#viewport").selectAll("g.nodes-container").data(pathway.nodes).enter().append("g").attr("id",function(d){return"nodes-container-"+d.graphId}).attr("transform",function(d){return"translate("+d.x+" "+d.y+")"}).attr("class","nodes-container").on("click",function(d){"data-node"===d.elementType&&pathvisio.pathway.xRef.displayData(pathway.organism,d)});nodesContainer.each(function(d,i){var node=d3.select(this).append("use").attr("id",function(d){return"node-"+d.graphId}).attr("transform",function(d){var transform="scale(1)";return d.hasOwnProperty("rotation")&&(transform="rotate("+d.rotation+" "+d.width/2+" "+d.height/2+")"),transform}).attr("class",function(d){var styleClass="";return styleClass="data-node"===d.elementType?"node "+d.elementType+" "+d.dataNodeType:"node "+d.elementType}).attr("x",0).attr("y",0).attr("width",function(d){return d.width}).attr("height",function(d){return d.height}).attr("z-index",function(d){return d.zIndex}).attr("style",function(d){var style="";d.hasOwnProperty("fill")&&(style+="fill:"+d.fill+"; "),d.hasOwnProperty("fillOpacity")&&(style+="fill-opacity:"+d.fillOpacity+"; "),d.hasOwnProperty("stroke")&&(style+="stroke:"+d.stroke+"; ");var strokeWidthEffective=null;return strokeWidthEffective=d.hasOwnProperty("strokeWidth")?2*d.strokeWidth:2,style+="stroke-width:"+strokeWidthEffective+"; ",d.hasOwnProperty("strokeStyle")&&("dashed"===d.strokeStyle&&(style+="stroke-dasharray: 5,3; "),"double"===d.strokeStyle&&d3.select(nodesContainer[0][i]).append("use").attr("id",function(d){return"node-double"+d.graphId}).attr("class",function(d){var styleClass="";return styleClass="data-node"===d.elementType?"node "+d.elementType+" "+d.dataNodeType:"node "+d.elementType}).attr("transform",function(d){var transform="scale(1)";return d.hasOwnProperty("rotation")&&(transform="rotate("+d.rotation+" "+d.width/2+" "+d.height/2+")"),transform}).attr("x",function(){return strokeWidthEffective}).attr("y",function(){return strokeWidthEffective}).attr("width",function(d){return d.width-2*strokeWidthEffective}).attr("height",function(d){return d.height-2*strokeWidthEffective}).attr("xlink:xlink:href",function(d){return"#"+d.symbolType}).attr("style",function(){return style+"fill-opacity:0; "})),style}),symbol=svg.select("symbol#"+d.symbolType);if(1===symbol.length?node.attr("xlink:xlink:href",function(d){return"#"+d.symbolType}):(node.attr("xlink:xlink:href","#rectangle"),console.warn("Pathvisio.js does not have access to the requested symbol: "+pathway.nodes[0].symbolType+". Rectangle used as placeholder.")),d.hasOwnProperty("textLabel")){var nodeText=d3.select(this).append("text").attr("id",function(d){return"node-text-"+d.graphId}).attr("x",0).attr("y",0).attr("transform",function(d){var dx=null;dx=d.textLabel.hasOwnProperty("textAnchor")?"start"===d.textLabel.textAnchor?5:"end"===d.textLabel.textAnchor?d.width-5:d.width/2:d.width/2;var dy=null;return dy=d.textLabel.hasOwnProperty("vAlign")?"top"===d.textLabel.vAlign?5+1*d.textLabel.fontSize:"bottom"===d.textLabel.vAlign?d.height-(5+.3*d.textLabel.fontSize+(pathvisio.helpers.splitStringByNewLine(d.textLabel.text).length-1)*d.textLabel.fontSize):d.height/2+.3*d.textLabel.fontSize-(pathvisio.helpers.splitStringByNewLine(d.textLabel.text).length-1)*d.textLabel.fontSize/2:d.height/2+.3*d.textLabel.fontSize-(pathvisio.helpers.splitStringByNewLine(d.textLabel.text).length-1)*d.textLabel.fontSize/2,"translate("+dx+" "+dy+")"}).attr("class",function(d){var styleClass="";return"data-node"===d.elementType&&(styleClass=d.dataNodeType),styleClass}).attr("style",function(d){var style="";return d.fontSize,d.textLabel.hasOwnProperty("fill")&&(style+="fill:"+d.textLabel.fill+"; "),d.textLabel.hasOwnProperty("fontFamily")&&(style+="font-family:"+d.textLabel.fontFamily+"; "),d.textLabel.hasOwnProperty("fontSize")&&(style+="font-size:"+d.textLabel.fontSize+"px; "),d.textLabel.hasOwnProperty("fontWeight")&&(style+="font-weight:"+d.textLabel.fontWeight+"; "),d.textLabel.hasOwnProperty("fontStyle")&&(style+="font-style:"+d.textLabel.fontStyle+"; "),d.textLabel.hasOwnProperty("textAnchor")&&(style+="text-anchor:"+d.textLabel.textAnchor+"; "),style});nodeText.each(function(d){var fontSize=d.textLabel.fontSize;d3.select(this).selectAll("tspan").data(function(d){var textArray=pathvisio.helpers.splitStringByNewLine(d.textLabel.text);return textArray}).enter().append("tspan").attr("x",0).attr("y",function(d,i){return i*fontSize}).text(function(d){return d})}),d.hasOwnProperty("biopaxRefs")&&d3.select(this).selectAll(".node-publication-xref-text").data(d.biopaxRefs).enter().append("text").attr("id",function(d){return"node-publication-xref-text-"+d}).attr("x",0).attr("y",0).attr("transform",function(d,i){return"translate("+12*i+" "+-12+")"}).attr("class","node-publication-xref-text").attr("style","").text(function(d){var index=0,rdfId=null;do console.log("pathway.biopax"),console.log(pathway.biopax),rdfId=pathway.biopax.bpPublicationXrefs[index].rdfId,index+=1;while(rdfId!==d.Text&&index<pathway.biopax.bpPublicationXrefs.length);return index})}})}function getPortCoordinates(boxDimensions,relX,relY){var port={};return port.x=boxDimensions.x+relX*boxDimensions.width,port.y=boxDimensions.y+relY*boxDimensions.height,port}var alignToAnchorMappings={Left:"start",Center:"middle",Right:"end"};return{drawAll:drawAll,getPortCoordinates:getPortCoordinates,gpml2json:gpml2json}}(),pathvisio.pathway.edge=function(){function gpml2json(rawJsonEdges){function attachedToAnchor(point,edges){var anchor=null,i=-1;do i+=1,anchor=edges[i].anchors.filter(function(element){return element.graphId===point.graphRef})[0];while(!anchor&&i<edges.length-1);return void 0!==anchor}try{rawJsonEdges.forEach(function(element){if(element.graphics.hasOwnProperty("anchor")&&(element.anchors=pathvisio.helpers.convertToArray(element.graphics.anchor)),element.graphics.hasOwnProperty("color")){var color=new RGBColor(element.graphics.color);color.ok&&(element.stroke=color.toHex())}element.strokeWidth=element.graphics.lineThickness,element.graphics.hasOwnProperty("connectorType")&&(element.connectorType=element.graphics.connectorType.toLowerCase()),element.graphics.hasOwnProperty("lineStyle")?(element.strokeStyle=element.graphics.lineStyle.toLowerCase(),"broken"===element.strokeStyle&&(element.strokeStyle="dashed")):element.hasOwnProperty("attribute")&&"org.pathvisio.DoubleLineProperty"===element.attribute.key&&"Double"===element.attribute.value&&(element.strokeStyle="double",delete element.attribute),element.zIndex=element.graphics.zorder,element.hasOwnProperty("xref")&&(element.xref.database||element.xref.id?(element.xref=element.xRef,delete element.xref):delete element.xref);var points=pathvisio.helpers.convertToArray(element.graphics.point),pointsData=pathvisio.pathway.edge.point.gpml2json(points);element.points=pointsData.points,element.markerStart=pointsData.markerStart,element.markerEnd=pointsData.markerEnd,delete element.graphics}),rawJsonEdges.sort(function(a,b){return a.zIndex-b.zIndex});var edgesWithAnchors=[],edgesWithoutAnchors=[];rawJsonEdges.forEach(function(element){element.hasOwnProperty("anchors")?edgesWithAnchors.push(element):edgesWithoutAnchors.push(element)});var validJsonEdges=[],unsortedJsonEdges=edgesWithAnchors;unsortedJsonEdges.forEach(function(element,index,array){attachedToAnchor(element.points[0],edgesWithAnchors)||attachedToAnchor(element.points[element.points.length-1],edgesWithAnchors)||(validJsonEdges.push(element),array.splice(index,1))});do unsortedJsonEdges.forEach(function(element,index,array){attachedToAnchor(element.points[0],edgesWithAnchors)&&!attachedToAnchor(element.points[0],validJsonEdges)||!attachedToAnchor(element.points[element.points.length-1],validJsonEdges)&&attachedToAnchor(element.points[element.points.length-1],edgesWithAnchors)||(validJsonEdges.push(element),array.splice(index,1))});while(unsortedJsonEdges.length>0);return validJsonEdges=validJsonEdges.concat(edgesWithoutAnchors)}catch(e){return console.log("Error converting edge to json: "+e.message),e}}function drawAll(svg,pathway){if(!svg||!pathway)return svg||console.log("svg"),pathway||console.log("pathway"),console.warn("Error: Missing one or more required parameters: svg, pathway.");if(pathway.hasOwnProperty("edges")){var pathData=null;svg.select("#viewport").selectAll("pathway.edge").data(pathway.edges).enter().append("path").attr("id",function(d){return d.edgeType+"-"+d.graphId}).attr("class",function(d){var styleClass="edge "+d.edgeType+" ";return d.hasOwnProperty("strokeStyle")&&"dashed"===d.strokeStyle&&(styleClass+=" dashed-stroke"),styleClass}).attr("style",function(d){var style="stroke-width:"+d.strokeWidth+"; ";return d.hasOwnProperty("stroke")&&(style+="stroke:"+d.stroke+"; "),d.hasOwnProperty("strokeStyle")&&"double"===d.strokeStyle&&(style+="stroke-width:"+3*d.strokeWidth+"; "),style}).attr("marker-start",function(d){var markerStart=pathvisio.pathway.edge.marker.draw(svg,d.markerStart,"start",d.stroke);return d.hasOwnProperty("strokeStyle")&&"double"===d.strokeStyle&&(markerStart="double-line-hack-start"),"url(#"+markerStart+")"}).attr("marker-end",function(d){var markerEnd=pathvisio.pathway.edge.marker.draw(svg,d.markerEnd,"end",d.stroke);return d.hasOwnProperty("strokeStyle")&&"double"===d.strokeStyle&&(markerEnd="double-line-hack-end"),"url(#"+markerEnd+")"}).attr("fill","none").attr("d",function(data){return console.log("prob svg"),console.log(svg),pathData=pathvisio.pathway.edge.pathData.get(svg,pathway,data),data.hasOwnProperty("strokeStyle")&&"double"===data.strokeStyle&&svg.select("#viewport").append("path").attr("class",data.edgeType+"-double").attr("d",pathData).attr("class","drawing-board-color-stroke").attr("style","stroke-width:"+data.strokeWidth+"; ").attr("marker-start","url(#"+pathvisio.pathway.edge.marker.draw(svg,data.markerStart,"start",data.stroke)+")").attr("marker-end","url(#"+pathvisio.pathway.edge.marker.draw(svg,data.markerEnd,"end",data.stroke)+")"),pathData})}}return{gpml2json:gpml2json,drawAll:drawAll}}(),pathvisio.pathway.edge.marker=function(){function draw(svg,name,position,color){var markerUrl="";if("none"===name)markerUrl=name;else{var markerElementBlack=svg.select("marker#"+name+"-"+position+"-black");if(1===markerElementBlack.length)if("#000"!==color&&"#000000"!==color&&color){var markerElement=pathvisio.helpers.cloneNode(markerElementBlack[0][0]),markerElementStyle="";"black"===markerElement[0][0].getAttribute("stroke")&&(markerElementStyle+="stroke:"+color+"; "),"black"===markerElement[0][0].getAttribute("fill")&&(markerElementStyle+="fill:"+color+"; "),markerElement[0][0].setAttribute("id",name+"-"+position+"-"+color),markerElement[0][0].setAttribute("style",markerElementStyle),markerUrl=name+"-"+position+"-"+color}else markerUrl=name+"-"+position+"-black";else markerUrl="none",console.warn("Pathvisio.js does not have access to the requested marker: "+name)}return markerUrl}return{draw:draw}}(),pathvisio.pathway.edge.point=function(){function gpml2json(rawJsonPoints){try{var markerStart="none",markerEnd="none";rawJsonPoints.forEach(function(element,index,array){if(void 0!==element.graphRef){delete element.x,delete element.y;var relX=(Math.round(2*element.relX)/2).toString();element.relX=parseFloat(anchorPositionMappings[relX]);var relY=(Math.round(2*element.relY)/2).toString();element.relY=parseFloat(anchorPositionMappings[relY]),0===element.relX?element.dx=-1:1===element.relX?element.dx=1:0===element.relY?element.dy=-1:1===element.relY&&(element.dy=1)}0===index&&markerMappings.hasOwnProperty(element.arrowHead)?(markerStart=markerMappings[element.arrowHead],delete element.arrowHead):index===array.length-1&&markerMappings.hasOwnProperty(element.arrowHead)&&(markerEnd=markerMappings[element.arrowHead],delete element.arrowHead)});var validJsonPoints=rawJsonPoints;return{points:validJsonPoints,markerStart:markerStart,markerEnd:markerEnd}}catch(e){return console.log("Error converting point to json: "+e.message),e}}function getGraphRef(pathway,point){if(point.hasOwnProperty("graphRef")){if(pathway.hasOwnProperty("nodes")){var node=pathway.nodes.filter(function(element){return element.graphId===point.graphRef})[0];if(void 0!==node)return{type:"node",element:node}}if(pathway.hasOwnProperty("groups")){var group=pathway.groups.filter(function(element){return element.graphId===point.graphRef})[0];if(void 0!==group)return{type:"group",groupId:group.groupId}}var edgesWithAnchors=pathway.edges.filter(function(element){return element.hasOwnProperty("anchors")}),anchor=null,i=-1;do i+=1,anchor=edgesWithAnchors[i].anchors.filter(function(element){return element.graphId===point.graphRef})[0];while(!anchor&&i<edgesWithAnchors.length);return{type:"anchor",element:anchor,edge:edgesWithAnchors[i]}}return{type:"unconnected"}}function getCoordinates(svg,pathway,point){if(!svg||!pathway||!point)return console.warn("Error: Missing input parameters.");var coordinates={},edgeTerminusRef=getGraphRef(pathway,point);
if("anchor"!==edgeTerminusRef.type)if("unconnected"===edgeTerminusRef.type)coordinates.x=point.x,coordinates.y=point.y;else if("node"===edgeTerminusRef.type)coordinates=pathvisio.pathway.node.getPortCoordinates(edgeTerminusRef.element,point.relX,point.relY);else{if("group"!==edgeTerminusRef.type)return"error";var groupDimensions=pathvisio.pathway.group.getDimensions(pathway,edgeTerminusRef.groupId);coordinates=pathvisio.pathway.node.getPortCoordinates(groupDimensions,point.relX,point.relY)}else{var path=svg.select("#interaction-"+edgeTerminusRef.edge.graphId)[0][0];coordinates=path.getPointAtLength(edgeTerminusRef.element.position*path.getTotalLength())}return coordinates}function isTwoPointElbow(source,target){var isRightAngle=Math.abs(source.dx)===Math.abs(target.dy)&&Math.abs(source.dy)===Math.abs(target.dx),sourcePasses=(target.x-source.x)/Math.abs(target.x-source.x)===source.dx||(target.y-source.y)/Math.abs(target.y-source.y)===source.dy,targetPasses=(source.x-target.x)/Math.abs(source.x-target.x)===target.dx||(source.y-target.y)/Math.abs(source.y-target.y)===target.dy;return isRightAngle&&sourcePasses&&targetPasses}var anchorPositionMappings={"-1":0,"-0.5":.333,0:.5,.5:.667,1:1},markerMappings={Arrow:"arrow","mim-branching-left":"mim-branching-left","mim-branching-right":"mim-branching-right","mim-necessary-stimulation":"mim-necessary-stimulation","mim-binding":"mim-binding","mim-conversion":"mim-conversion","mim-stimulation":"mim-stimulation","mim-modification":"mim-modification","mim-catalysis":"mim-catalysis","mim-inhibition":"mim-inhibition","mim-cleavage":"mim-cleavage","mim-covalent-bond":"mim-covalent-bond","mim-transcription-translation":"mim-transcription-translation","mim-gap":"mim-gap",none:"none",TBar:"t-bar"};return{getGraphRef:getGraphRef,getCoordinates:getCoordinates,isTwoPointElbow:isTwoPointElbow,gpml2json:gpml2json}}(),pathvisio.pathway.edge.pathData=function(){function switchDirection(currentDirection){return"H"===currentDirection?"V":"H"}function get(svg,pathway,edge){if(!svg||!edge)return console.warn("Error: Missing input parameters.");var sourcePoint=edge.points[0],source=pathvisio.pathway.edge.point.getCoordinates(svg,pathway,sourcePoint);source.dx=void 0===sourcePoint.dx?0:sourcePoint.dx,source.dy=void 0===sourcePoint.dy?0:sourcePoint.dy;var targetPoint=edge.points[edge.points.length-1],target=pathvisio.pathway.edge.point.getCoordinates(svg,pathway,targetPoint);target.dx=void 0===targetPoint.dx?0:targetPoint.dx,target.dy=void 0===targetPoint.dy?0:targetPoint.dy;var pathData="M "+source.x+" "+source.y;if(edge.connectorType&&void 0!==edge.connectorType&&"straight"!==edge.connectorType)if("elbow"===edge.connectorType&&edge.points[0].hasOwnProperty("graphRef")&&edge.points[edge.points.length-1].hasOwnProperty("graphRef"))currentDirection=1===Math.abs(source.dx)?"H":"V",2===edge.points.length&&pathvisio.pathway.edge.point.isTwoPointElbow(source,target)||(edge.points.length>2?edge.points.forEach(function(element,index,array){index>0&&index<array.length-1&&(pathData+="H"===currentDirection?" "+currentDirection+" "+element.x:" "+currentDirection+" "+element.y,currentDirection=switchDirection(currentDirection))}):(1===Math.abs(source.dx)?pathData+=" H "+(source.x+15*source.dx):1===Math.abs(source.dy)&&(pathData+=" V "+(source.y+15*source.dy),currentDirection=switchDirection(currentDirection)),target.dx===(target.x-source.x)/Math.abs(target.x-source.x)||source.dx===target.dy||source.dy===target.dx?(pathData+=" H "+(target.x+15*target.dx)+" V "+target.y+" H "+target.x,currentDirection=switchDirection(currentDirection)):(target.dy===(target.y-source.y)/Math.abs(target.y-source.y)||source.dx===target.dy||source.dy===target.dx)&&(pathData+=" V "+(target.y+15*target.dy)+" H "+target.x+" V "+target.y,currentDirection=switchDirection(currentDirection)))),"H"===currentDirection?(pathData+=" "+currentDirection+" "+target.x,currentDirection=switchDirection(currentDirection),pathData+=" "+currentDirection+" "+target.y,currentDirection=switchDirection(currentDirection)):(pathData+=" "+currentDirection+" "+target.y,currentDirection=switchDirection(currentDirection),pathData+=" "+currentDirection+" "+target.x,currentDirection=switchDirection(currentDirection));else if("segmented"===edge.connectorType)edge.points.forEach(function(element,index,array){index>0&&index<array.length-1&&(pathData+=" L "+element.x+" "+element.y)}),pathData+=" L "+target.x+" "+target.y;else{if("curved"===edge.connectorType){if(3===edge.points.length){var pointControl=edge.points[1];return pathData+=" S"+pointControl.x+","+pointControl.y+" "+target.x+","+target.y}return pathData+=" T"+target.x+","+target.y}console.log("Warning: pathvisio.js does not support connector type: "+edge.connectorType),edge.points.forEach(function(element,index,array){index>0&&index<array.length-1&&(pathData+=" L "+element.x+" "+element.y)}),pathData+=" L "+target.x+" "+target.y}else pathData+=" L "+target.x+" "+target.y;return pathData}var currentDirection=null;return{get:get}}(),pathvisio.pathway.dataSources=[{database:"Affy",id:"X",homePage:"http://www.affymetrix.com/",linkOut:"https://www.affymetrix.com/LinkServlet?probeset=$id",example:"1851_s_at",dataNodeType:"probe",species:"",priority:0,unknown:"urn:miriam:affy.probeset",regex:"d{4,}((_[asx])?_at)?",fullName:"Affymetrix Probeset"},{database:"Agilent",id:"Ag",homePage:"http://agilent.com",linkOut:"",example:"A_24_P98555",dataNodeType:"probe",species:"",priority:0,unknown:"Ag",regex:"A_d+_.+",fullName:"Agilent"},{database:"BIND",id:"Bi",homePage:"http://www.bind.ca/",linkOut:"http://www.bind.ca/Action?identifier=bindid&idsearch=$id",example:"",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:bind",regex:"^d+$",fullName:"BIND"},{database:"BioCyc",id:"Bc",homePage:"http://biocyc.org",linkOut:"http://biocyc.org/getid?id=$id",example:"ECOLI:CYT-D-UBIOX-CPLX",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:biocyc",regex:"^w+:[A-Za-z0-9-]+$",fullName:"BioCyc"},{database:"BioGrid",id:"Bg",homePage:"http://thebiogrid.org/",linkOut:"http://thebiogrid.org/$id",example:"31623",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:biogrid",regex:"^d+$",fullName:"BioGRID"},{database:"BioModels Database",id:"Bm",homePage:"http://www.ebi.ac.uk/biomodels/",linkOut:"http://www.ebi.ac.uk/biomodels-main/$id",example:"BIOMD0000000048",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:biomodels.db",regex:"^((BIOMD|MODEL)d{10})|(BMIDd{12})$",fullName:"BioModels Database"},{database:"BioSystems",id:"Bs",homePage:"http://www.ncbi.nlm.nih.gov/biosystems/",linkOut:"http://www.ncbi.nlm.nih.gov/biosystems/$id",example:"1",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:biosystems",regex:"^d+$",fullName:"BioSystems"},{database:"BRENDA",id:"Br",homePage:"http://www.brenda-enzymes.org/",linkOut:"http://www.brenda-enzymes.org/php/result_flat.php4?ecno=$id",example:"1.1.1.1",dataNodeType:"",species:"",priority:1,unknown:"urn:miriam:brenda",regex:"^((d+.-.-.-)|(d+.d+.-.-)|(d+.d+.d+.-)|(d+.d+.d+.d+))$",fullName:"BRENDA"},{database:"CAS",id:"Ca",homePage:"http://commonchemistry.org",linkOut:"http://commonchemistry.org/ChemicalDetail.aspx?ref=$id",example:"50-00-0",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:cas",regex:"^d{1,7}-d{2}-d$",fullName:"CAS"},{database:"CCDS",id:"Cc",homePage:"http://identifiers.org/ccds/",linkOut:"http://www.ncbi.nlm.nih.gov/CCDS/CcdsBrowse.cgi?REQUEST=ALLFIELDS&DATA=$id",example:"CCDS33337",dataNodeType:"gene",species:"",priority:0,unknown:"",regex:"^CCDSd+.d+$",fullName:"Consensus CDS"},{database:"ChEBI",id:"Ce",homePage:"http://www.ebi.ac.uk/chebi/",linkOut:"http://www.ebi.ac.uk/chebi/searchId.do?chebiId=$id",example:"CHEBI:36927",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:obo.chebi",regex:"^CHEBI:d+$",fullName:"ChEBI"},{database:"Chemspider",id:"Cs",homePage:"http://www.chemspider.com/",linkOut:"http://www.chemspider.com/Chemical-Structure.$id.html",example:"56586",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:chemspider",regex:"^d+$",fullName:"ChemSpider"},{database:"CodeLink",id:"Ge",homePage:"http://www.appliedmicroarrays.com/",linkOut:"",example:"GE86325",dataNodeType:"probe",species:"",priority:0,unknown:"Ge",regex:"",fullName:"CodeLink"},{database:"Database of Interacting Proteins",id:"Dip",homePage:"http://dip.doe-mbi.ucla.edu/",linkOut:"http://dip.doe-mbi.ucla.edu/dip/DIPview.cgi?ID=$id",example:"DIP-743N",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:dip",regex:"^DIP[:-]d{3}[EN]$",fullName:"Database of Interacting Proteins"},{database:"dbSNP",id:"Sn",homePage:"http://www.ncbi.nlm.nih.gov/sites/entrez?db=snp",linkOut:"http://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=$id",example:"121909098",dataNodeType:"gene",species:"",priority:1,unknown:"",regex:"^d+$",fullName:"dbSNP"},{database:"DrugBank",id:"Dr",homePage:"http://www.drugbank.ca/",linkOut:"http://www.drugbank.ca/drugs/$id",example:"DB00001",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:drugbank",regex:"^DBd{5}$",fullName:"DrugBank"},{database:"EcoCyc",id:"Eco",homePage:"http://ecocyc.org/",linkOut:"http://ecocyc.org/ECOLI/NEW-IMAGE?type=NIL&object=$id",example:"325-BISPHOSPHATE-NUCLEOTIDASE-RXN",dataNodeType:"interaction",species:"Escherichia coli",priority:1,unknown:"Eco",regex:"",fullName:"EcoCyc"},{database:"EcoGene",id:"Ec",homePage:"http://ecogene.org/",linkOut:"http://ecogene.org/geneInfo.php?eg_id=$id",example:"EG10173",dataNodeType:"gene",species:"Escherichia coli",priority:1,unknown:"urn:miriam:ecogene",regex:"^EGd+$",fullName:"EcoGene"},{database:"EMBL",id:"Em",homePage:"http://www.ebi.ac.uk/embl/",linkOut:"http://www.ebi.ac.uk/ena/data/view/$id",example:"X58356",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:ena.embl",regex:"^[A-Z]+[0-9]+$",fullName:"European Nucleotide Archive"},{database:"Ensembl",id:"En",homePage:"http://www.ensembl.org/",linkOut:"http://www.ensembl.org/id/$id",example:"ENSG00000139618",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:ensembl",regex:"^ENS[A-Z]*[FPTG]d{11}$",fullName:"Ensembl"},{database:"Ensembl B. subtilis",id:"EnBs",homePage:"http://www.ensembl.org",linkOut:"http://bacteria.ensembl.org/Bacillus/B_subtilis/Gene/Summary?g=$id",example:"EBBACG00000000013",dataNodeType:"gene",species:"Bacillus subtilis",priority:1,unknown:"EnBs",regex:"EBBACGd{11}",fullName:"Ensembl"},{database:"Ensembl C. elegans",id:"EnCe",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Caenorhabditis_elegans/Gene/Summary?g=$id",example:"Y42H9B.1",dataNodeType:"gene",species:"Caenorhabditis elegans",priority:1,unknown:"EnCe",regex:"",fullName:"Ensembl"},{database:"Ensembl Chicken",id:"EnGg",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Gallus_gallus/Gene/Summary?g=$id",example:"ENSGALG00000021736",dataNodeType:"gene",species:"Gallus gallus",priority:1,unknown:"EnGg",regex:"ENSGALGd{11}",fullName:"Ensembl"},{database:"Ensembl Chimp",id:"EnPt",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Pan_troglodytes/Gene/Summary?g=$id",example:"ENSPTRG00000036034",dataNodeType:"gene",species:"Pan troglodytes",priority:1,unknown:"EnPt",regex:"ENSPTRGd{11}",fullName:"Ensembl"},{database:"Ensembl Cow",id:"EnBt",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Bos_taurus/Gene/Summary?g=$id",example:"ENSBTAG00000043548",dataNodeType:"gene",species:"Bos taurus",priority:1,unknown:"EnBt",regex:"ENSBTAGd{11}",fullName:"Ensembl"},{database:"Ensembl Dog",id:"EnCf",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Canis_familiaris/Gene/Summary?g=$id",example:"ENSCAFG00000025860",dataNodeType:"gene",species:"Canis familiaris",priority:1,unknown:"EnCf",regex:"ENSCAFGd{11}",fullName:"Ensembl"},{database:"Ensembl E. coli",id:"EnEc",homePage:"http://www.ensembl.org",linkOut:"http://bacteria.ensembl.org/Escherichia_Shigella/E_coli_K12/Gene/Summary?g=$id",example:"EBESCG00000000010",dataNodeType:"gene",species:"Escherichia coli",priority:1,unknown:"EnEc",regex:"EBESCGd{11}",fullName:"Ensembl"},{database:"Ensembl Fruitfly",id:"EnDm",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Drosophila_melanogaster/Gene/Summary?g=$id",example:"FBgn0032956",dataNodeType:"gene",species:"Drosophila melanogaster",priority:1,unknown:"EnDm",regex:"FBgnd{7}",fullName:"Ensembl"},{database:"Ensembl Horse",id:"EnQc",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Equus_caballus/Gene/Summary?g=$id",example:"ENSECAG00000026160",dataNodeType:"gene",species:"Equus caballus",priority:1,unknown:"EnQc",regex:"ENSECAGd{11}",fullName:"Ensembl"},{database:"Ensembl Human",id:"EnHs",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=$id",example:"ENSG00000139618",dataNodeType:"gene",species:"Homo sapiens",priority:1,unknown:"EnHs",regex:"ENSGd{11}",fullName:"Ensembl"},{database:"Ensembl M. tuberculosis",id:"EnMx",homePage:"http://www.ensembl.org",linkOut:"http://bacteria.ensembl.org/Mycobacterium/M_tuberculosis_H37Rv/Gene/Summary?g=$id",example:"EBMYCG00000003122",dataNodeType:"gene",species:"Mycobacterium tuberculosis",priority:1,unknown:"EnMx",regex:"EBMYCGd{11}",fullName:"Ensembl"},{database:"Ensembl Mosquito",id:"EnAg",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Anopheles_gambiae/Gene/Summary?_q=$id",example:"AGAP006864",dataNodeType:"gene",species:"Anopheles gambiae",priority:1,unknown:"EnAg",regex:"AGAPd{6}",fullName:"Ensembl"},{database:"Ensembl Mouse",id:"EnMm",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Mus_musculus/Gene/Summary?g=$id",example:"ENSMUSG00000017167",dataNodeType:"gene",species:"Mus musculus",priority:1,unknown:"EnMm",regex:"ENSMUSGd{11}",fullName:"Ensembl"},{database:"Ensembl Pig",id:"EnSs",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Sus_scrofa/Gene/Summary?g=$id",example:"ENSSSCG00000004244",dataNodeType:"gene",species:"Sus scrofa",priority:1,unknown:"EnSs",regex:"ENSSSCGd{11}",fullName:"Ensembl"},{database:"Ensembl Plants",id:"EP",homePage:"http://plants.ensembl.org/",linkOut:"http://plants.ensembl.org/id/$id",example:"AT1G73965",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:ensembl.plant",regex:"^w+$",fullName:"Ensembl Plants"},{database:"Ensembl Rat",id:"EnRn",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Rattus_norvegicus/Gene/Summary?g=$id",example:"ENSRNOG00000016648",dataNodeType:"gene",species:"Rattus norvegicus",priority:1,unknown:"EnRn",regex:"ENSRNOGd{11}",fullName:"Ensembl"},{database:"Ensembl Xenopus",id:"EnXt",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Xenopus_tropicalis/Gene/Summary?g=$id",example:"ENSXETG00000029448",dataNodeType:"gene",species:"Xenopus tropicalis",priority:1,unknown:"EnXt",regex:"ENSXETGd{11}",fullName:"Ensembl"},{database:"Ensembl Yeast",id:"EnSc",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Saccharomyces_cerevisiae/Gene/Summary?g=$id",example:"YGR147C",dataNodeType:"gene",species:"Saccharomyces cerevisiae",priority:1,unknown:"EnSc",regex:"Y[A-Z][RL]d{3}[WC](?:-[A-Z])?",fullName:"Ensembl"},{database:"Ensembl Zebrafish",id:"EnDr",homePage:"http://www.ensembl.org",linkOut:"http://www.ensembl.org/Danio_rerio/Gene/Summary?g=$id",example:"ENSDARG00000024771",dataNodeType:"gene",species:"Danio rerio",priority:1,unknown:"EnDr",regex:"ENSDARGd{11}",fullName:"Ensembl"},{database:"Entrez Gene",id:"L",homePage:"http://www.ncbi.nlm.nih.gov/gene",linkOut:"http://www.ncbi.nlm.nih.gov/gene/$id",example:"100010",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:ncbigene",regex:"^d+$",fullName:"Entrez Gene"},{database:"Enzyme Nomenclature",id:"E",homePage:"http://www.ebi.ac.uk/intenz/",linkOut:"http://www.ebi.ac.uk/intenz/query?cmd=SearchEC&ec=$id",example:"1.1.1.1",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:ec-code",regex:"^d+.-.-.-|d+.d+.-.-|d+.d+.d+.-|d+.d+.d+.(n)?d+$",fullName:"Enzyme Nomenclature"},{database:"FlyBase",id:"F",homePage:"http://flybase.org/",linkOut:"http://flybase.org/reports/$id.html",example:"FBgn0011293",dataNodeType:"gene",species:"Drosophila melanogaster",priority:1,unknown:"urn:miriam:flybase",regex:"^FBw{2}d{7}$",fullName:"FlyBase"},{database:"GenBank",id:"G",homePage:"http://www.ncbi.nlm.nih.gov/genbank/",linkOut:"http://www.ncbi.nlm.nih.gov/nuccore/$id",example:"NW_004190323",dataNodeType:"gene",species:"",priority:0,unknown:"G",regex:"(wd{5})|(w{2}d{6})|(w{3}d{5})",fullName:"GenBank"},{database:"Gene Wiki",id:"Gw",homePage:"http://en.wikipedia.org/wiki/Portal:Gene_Wiki",linkOut:"http://plugins.biogps.org/cgi-bin/wp.cgi?id=$id",example:"1017",dataNodeType:"gene",species:"",priority:0,unknown:"Gw",regex:"d+",fullName:"Gene Wiki"},{database:"GeneOntology",id:"T",homePage:"http://www.ebi.ac.uk/QuickGO/",linkOut:"http://www.ebi.ac.uk/QuickGO/GTerm?id=$id",example:"GO:0006915",dataNodeType:"gene",species:"",priority:0,unknown:"urn:miriam:obo.go",regex:"^GO:d{7}$",fullName:"Gene Ontology"},{database:"Gramene Arabidopsis",id:"EnAt",homePage:"http://www.gramene.org/",linkOut:"http://www.gramene.org/Arabidopsis_thaliana/Gene/Summary?g=$id",example:"ATMG01360-TAIR-G",dataNodeType:"gene",species:"Arabidopsis thaliana",priority:1,unknown:"EnAt",regex:"AT[dCM]Gd{5}-TAIR-G",fullName:"Grameen Arabidopsis"},{database:"Gramene Genes DB",id:"Gg",homePage:"http://www.gramene.org/",linkOut:"http://www.gramene.org/db/genes/search_gene?acc=$id",example:"GR:0060184",dataNodeType:"gene",species:"",priority:1,unknown:"Gg",regex:"GR:d+",fullName:"Gramene Genes"},{database:"Gramene Literature",id:"Gl",homePage:"http://www.gramene.org/",linkOut:"http://www.gramene.org/db/literature/pub_search?ref_id=$id",example:"6200",dataNodeType:"gene",species:"",priority:0,unknown:"Gl",regex:"",fullName:"Gramene Literature"},{database:"Gramene Maize",id:"EnZm",homePage:"http://www.ensembl.org",linkOut:"http://www.maizesequence.org/Zea_mays/Gene/Summary?g=$id",example:"GRMZM2G174107",dataNodeType:"gene",species:"",priority:1,unknown:"EnZm",regex:"",fullName:"Gramene Maize"},{database:"Gramene Pathway",id:"Gp",homePage:"http://www.gramene.org/pathway",linkOut:"",example:"AAH72400",dataNodeType:"pathway",species:"",priority:1,unknown:"Gp",regex:"",fullName:"Gramene Pathway"},{database:"Gramene Rice",id:"EnOj",homePage:"http://www.gramene.org/",linkOut:"http://www.gramene.org/Oryza_sativa/Gene/Summary?db=core;g=$id",example:"osa-MIR171a",dataNodeType:"gene",species:"",priority:1,unknown:"EnOj",regex:"",fullName:"Gramene Rice"},{database:"HGNC",id:"H",homePage:"http://www.genenames.org",linkOut:"http://www.genenames.org/data/hgnc_data.php?match=$id",example:"DAPK1",dataNodeType:"gene",species:"Homo sapiens",priority:1,unknown:"urn:miriam:hgnc.symbol",regex:"^[A-Za-z0-9]+",fullName:"HGNC Symbol"},{database:"HGNC Accession number",id:"Hac",homePage:"http://www.genenames.org",linkOut:"http://www.genenames.org/data/hgnc_data.php?hgnc_id=$id",example:"HGNC:2674",dataNodeType:"gene",species:"Homo sapiens",priority:1,unknown:"urn:miriam:hgnc",regex:"^(HGNC:)?d{1,5}$",fullName:"HGNC"},{database:"HMDB",id:"Ch",homePage:"http://www.hmdb.ca/",linkOut:"http://www.hmdb.ca/metabolites/$id",example:"HMDB00001",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:hmdb",regex:"^HMDBd{5}$",fullName:"HMDB"},{database:"HomoloGene",id:"Hg",homePage:"http://www.ncbi.nlm.nih.gov/homologene/",linkOut:"http://www.ncbi.nlm.nih.gov/homologene/$id",example:"1000",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:homologene",regex:"^d+$",fullName:"HomoloGene"},{database:"HPRD",id:"Hp",homePage:"http://www.hprd.org/",linkOut:"",example:"",dataNodeType:"interaction",species:"Homo sapiens",priority:1,unknown:"urn:miriam:hprd",regex:"",fullName:"HPRD"},{database:"Illumina",id:"Il",homePage:"http://www.illumina.com/",linkOut:"",example:"ILMN_5668",dataNodeType:"probe",species:"",priority:0,unknown:"Il",regex:"ILMN_d+",fullName:"Illumina"},{database:"IntAct",id:"Ia",homePage:"http://www.ebi.ac.uk/intact/",linkOut:"http://www.ebi.ac.uk/intact/pages/details/details.xhtml?interactionAc=$id",example:"EBI-2307691",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:intact",regex:"^EBI-[0-9]+$",fullName:"IntAct"},{database:"InterPro",id:"I",homePage:"http://www.ebi.ac.uk/interpro/",linkOut:"http://www.ebi.ac.uk/interpro/DisplayIproEntry?ac=$id",example:"IPR000100",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:interpro",regex:"^IPRd{6}$",fullName:"InterPro"},{database:"IPI",id:"Ip",homePage:"http://www.ebi.ac.uk/IPI",linkOut:"http://www.ebi.ac.uk/cgi-bin/dbfetch?db=IPI&id=$id&format=default",example:"IPI00000001",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:ipi",regex:"^IPId{8}$",fullName:"IPI"},{database:"IRGSP Gene",id:"Ir",homePage:"http://rgp.dna.affrc.go.jp/IRGSP/",linkOut:"",example:"Os12g0561000",dataNodeType:"gene",species:"",priority:1,unknown:"Ir",regex:"Osd{2}gd+",fullName:"IRGSP Gene"},{database:"Kegg Compound",id:"Ck",homePage:"http://www.genome.jp/kegg/ligand.html",linkOut:"http://www.genome.jp/dbget-bin/www_bget?cpd:$id",example:"C12345",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:kegg.compound",regex:"^Cd+$",fullName:"KEGG Compound"},{database:"KEGG Drug",id:"Kd",homePage:"http://www.genome.jp/kegg/drug/",linkOut:"http://www.genome.jp/dbget-bin/www_bget?dr:$id",example:"D00123",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:kegg.drug",regex:"^Dd+$",fullName:"KEGG Drug"},{database:"KEGG Genes",id:"Kg",homePage:"http://www.genome.jp/kegg/genes.html",linkOut:"http://www.genome.jp/dbget-bin/www_bget?$id",example:"syn:ssr3451",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:kegg.genes",regex:"^w+:[wd.-]*$",fullName:"KEGG Genes"},{database:"KEGG Glycan",id:"Kgl",homePage:"http://www.genome.jp/kegg/glycan/",linkOut:"http://www.genome.jp/dbget-bin/www_bget?gl:$id",example:"G00123",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:kegg.glycan",regex:"^Gd+$",fullName:"KEGG Glycan"},{database:"KEGG Pathway",id:"Kp",homePage:"http://www.genome.jp/kegg/pathway.html",linkOut:"http://www.genome.jp/dbget-bin/www_bget?pathway+$id",example:"hsa00620",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:kegg.pathway",regex:"^w{2,4}d{5}$",fullName:"KEGG Pathway"},{database:"KEGG Reaction",id:"Kr",homePage:"http://www.genome.jp/kegg/reaction/",linkOut:"http://www.genome.jp/dbget-bin/www_bget?rn:$id",example:"R00100",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:kegg.reaction",regex:"^Rd+$",fullName:"KEGG Reaction"},{database:"LIPID MAPS",id:"Lm",homePage:"http://www.lipidmaps.org",linkOut:"http://www.lipidmaps.org/data/get_lm_lipids_dbgif.php?LM_ID=$id",example:"LMPR0102010012",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:lipidmaps",regex:"^LM(FA|GL|GP|SP|ST|PR|SL|PK)[0-9]{4}([0-9a-zA-Z]{4,6})?$",fullName:"LIPID MAPS"},{database:"LipidBank",id:"Lb",homePage:"http://lipidbank.jp/index.html",linkOut:"http://lipidbank.jp/cgi-bin/detail.cgi?id=$id",example:"BBA0001",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:lipidbank",regex:"^w+d+$",fullName:"LipidBank"},{database:"MACiE",id:"Ma",homePage:"http://www.ebi.ac.uk/thornton-srv/databases/MACiE/index.html",linkOut:"http://www.ebi.ac.uk/thornton-srv/databases/cgi-bin/MACiE/entry/getPage.pl?id=$id",example:"M0001",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:macie",regex:"^Md{4}$",fullName:"MACiE"},{database:"MaizeGDB",id:"Mg",homePage:"",linkOut:"http://www.maizegdb.org/cgi-bin/displaylocusresults.cgi?term=$id",example:"acc1",dataNodeType:"gene",species:"Zea mays",priority:1,unknown:"Mg",regex:"",fullName:"MaizeGDB"},{database:"MatrixDB",id:"Md",homePage:"http://matrixdb.ibcp.fr/",linkOut:"http://matrixdb.ibcp.fr/cgi-bin/model/report/default?name=$id&class=Association",example:"P00747_P07355",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:matrixdb.association",regex:"^([A-N,R-Z][0-9][A-Z][A-Z, 0-9][A-Z, 0-9][0-9])_.*|([O,P,Q][0-9][A-Z, 0-9][A-Z, 0-9][A-Z, 0-9][0-9]_.*)|(GAG_.*)|(MULT_.*)|(PFRAG_.*)|(LIP_.*)|(CAT_.*)$",fullName:"MatrixDB"},{database:"MetaCyc",id:"Mc",homePage:"http://www.metacyc.org/",linkOut:"http://www.metacyc.org/META/NEW-IMAGE?type=NIL&object=$id",example:"D-GLUTAMATE-OXIDASE-RXN",dataNodeType:"interaction",species:"",priority:1,unknown:"Mc",regex:"",fullName:"MetaCyc"},{database:"MGI",id:"M",homePage:"http://www.informatics.jax.org/",linkOut:"http://www.informatics.jax.org/marker/$id",example:"MGI:2442292",dataNodeType:"gene",species:"Mus musculus",priority:1,unknown:"urn:miriam:mgd",regex:"^MGI:d+$",fullName:"Mouse Genome Database"},{database:"MINT",id:"Mi",homePage:"http://mint.bio.uniroma2.it/mint/",linkOut:"http://mint.bio.uniroma2.it/mint/search/inFrameInteraction.do?interactionAc=$id",example:"MINT-10000",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:mint",regex:"^MINT-d{1,5}$",fullName:"MINT"},{database:"miRBase mature sequence",id:"Mbm",homePage:"http://www.mirbase.org/",linkOut:"http://www.mirbase.org/cgi-bin/mature.pl?mature_acc=$id",example:"MIMAT0000001",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:mirbase.mature",regex:"MIMATd{7}",fullName:"miRBase mature sequence"},{database:"miRBase Sequence",id:"Mb",homePage:"http://microrna.sanger.ac.uk/",linkOut:"http://microrna.sanger.ac.uk/cgi-bin/sequences/mirna_entry.pl?acc=$id",example:"MI0000001",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:mirbase",regex:"MId{7}",fullName:"miRBase Sequence"},{database:"NASC Gene",id:"N",homePage:"http://arabidopsis.info/",linkOut:"",example:"ATMG00960-TAIR-G",dataNodeType:"gene",species:"Arabidopsis thaliana",priority:1,unknown:"N",regex:"AT[dCM]Gd{5}-TAIR-G",fullName:"NASC Gene"},{database:"NCBI Protein",id:"Np",homePage:"http://www.ncbi.nlm.nih.gov/protein",linkOut:"http://www.ncbi.nlm.nih.gov/protein/$id",example:"CAA71118.1",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:ncbiprotein",regex:"^w+d+(.d+)?$",fullName:"NCBI Protein"},{database:"NCI Pathway Interaction Database",id:"Pid",homePage:"http://pid.nci.nih.gov/",linkOut:"http://pid.nci.nih.gov/search/pathway_landing.shtml?what=graphic&jpg=on&pathway_id=$id",example:"pi3kcipathway",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:pid.pathway",regex:"^w+$",fullName:"NCI Pathway Interaction Database"},{database:"NuGO wiki",id:"Nw",homePage:"http://wiki.nugo.org",linkOut:"http://wiki.nugo.org/index.php/$id",example:"HMDB00001",dataNodeType:"metabolite",species:"",priority:0,unknown:"Nw",regex:"",fullName:"NuGO wiki"},{database:"OMIM",id:"Om",homePage:"http://omim.org/",linkOut:"http://omim.org/entry/$id",example:"603903",dataNodeType:"gene",species:"",priority:0,unknown:"urn:miriam:omim",regex:"^[*#+%^]?d{6}$",fullName:"OMIM"},{database:"Oryzabase",id:"Ob",homePage:"http://www.shigen.nig.ac.jp/rice/oryzabase",linkOut:"http://www.shigen.nig.ac.jp/rice/oryzabase/gateway/gatewayAction.do?target=symbol&id=$id",example:"468",dataNodeType:"gene",species:"Oryza sativa",priority:1,unknown:"Ob",regex:"",fullName:"Oryzabase"},{database:"Other",id:"O",homePage:"",linkOut:"",example:"",dataNodeType:"gene",species:"",priority:1,unknown:"O",regex:"",fullName:"Other"},{database:"Pathway Commons",id:"Pc",homePage:"http://www.pathwaycommons.org/pc/",linkOut:"http://www.pathwaycommons.org/pc/record2.do?id=$id",example:"485991",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:pathwaycommons",regex:"^d+$",fullName:"Pathway Commons"},{database:"PDB",id:"Pd",homePage:"http://www.pdb.org/",linkOut:"http://www.rcsb.org/pdb/explore/explore.do?structureId=$id",example:"2gc4",dataNodeType:"protein",species:"",priority:0,unknown:"urn:miriam:pdb",regex:"^[0-9][A-Za-z0-9]{3}$",fullName:"Protein Data Bank"},{database:"Pfam",id:"Pf",homePage:"http://pfam.sanger.ac.uk/",linkOut:"http://pfam.sanger.ac.uk/family/$id/",example:"PF01234",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:pfam",regex:"^PFd{5}$",fullName:"Pfam"},{database:"PharmGKB Drug",id:"Pgd",homePage:"http://www.pharmgkb.org/",linkOut:"http://www.pharmgkb.org/drug/$id",example:"PA448710",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:pharmgkb.drug",regex:"^PAd+$",fullName:"PharmGKB Drug"},{database:"PharmGKB Gene",id:"Pgg",homePage:"http://www.pharmgkb.org/",linkOut:"http://www.pharmgkb.org/gene/$id",example:"PA131",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:pharmgkb.gene",regex:"^PAw+$",fullName:"PharmGKB Gene"},{database:"PharmGKB Pathways",id:"Pgp",homePage:"http://www.pharmgkb.org/",linkOut:"http://www.pharmgkb.org/pathway/$id",example:"PA146123006",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:pharmgkb.pathways",regex:"^PAd+$",fullName:"PharmGKB Pathways"},{database:"PhosphoSite Protein",id:"Pp",homePage:"http://www.phosphosite.org/homeAction.do",linkOut:"http://www.phosphosite.org/proteinAction.do?id=$id",example:"12300",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:phosphosite.protein",regex:"^d{5}$",fullName:"PhosphoSite Protein"},{database:"PINA",id:"Pi",homePage:"http://cbg.garvan.unsw.edu.au/pina/",linkOut:"http://cbg.garvan.unsw.edu.au/pina/interactome.oneP.do?ac=$id&showExtend=null",example:"Q13485",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:pina",regex:"^([A-N,R-Z][0-9][A-Z][A-Z, 0-9][A-Z, 0-9][0-9])|([O,P,Q][0-9][A-Z, 0-9][A-Z, 0-9][A-Z, 0-9][0-9])$",fullName:"PINA"},{database:"PlantGDB",id:"Pl",homePage:"http://www.plantgdb.org/",linkOut:"",example:"PUT-157a-Vitis_vinifera-37378",dataNodeType:"gene",species:"",priority:1,unknown:"Pl",regex:"PUT-[wd-]+",fullName:"PlantGDB"},{database:"PubChem-bioassay",id:"Cpb",homePage:"http://www.ncbi.nlm.nih.gov/sites/entrez?db=pcassay ",linkOut:"http://pubchem.ncbi.nlm.nih.gov/assay/assay.cgi?aid=$id",example:"1018",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:pubchem.bioassay",regex:"^d+$",fullName:"PubChem-bioassay"},{database:"PubChem-compound",id:"Cpc",homePage:"http://pubchem.ncbi.nlm.nih.gov/",linkOut:"http://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi?cid=$id",example:"100101",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:pubchem.compound",regex:"^d+$",fullName:"PubChem-compound"},{database:"PubChem-substance",id:"Cps",homePage:"http://pubchem.ncbi.nlm.nih.gov/",linkOut:"http://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi?sid=$id",example:"100101",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:pubchem.substance",regex:"^d+$",fullName:"PubChem-substance"},{database:"Reactome",id:"Re",homePage:"http://www.reactome.org/",linkOut:"http://www.reactome.org/cgi-bin/eventbrowser_st_id?FROM_REACTOME=1&ST_ID=$id",example:"REACT_1590",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:reactome",regex:"^REACT_d+(.d+)?$",fullName:"Reactome"},{database:"RefSeq",id:"Q",homePage:"http://www.ncbi.nlm.nih.gov/projects/RefSeq/",linkOut:"http://www.ncbi.nlm.nih.gov/entrez/viewer.fcgi?val=$id",example:"NP_012345",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:refseq",regex:"^(NC|AC|NG|NT|NW|NZ|NM|NR|XM|XR|NP|AP|XP|ZP)_d+$",fullName:"RefSeq"},{database:"RESID",id:"Res",homePage:"http://www.ebi.ac.uk/RESID/",linkOut:"http://srs.ebi.ac.uk/srsbin/cgi-bin/wgetz?-id+6JSUg1NA6u4+-e+[RESID:'$id']",example:"AA0001",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:resid",regex:"^AAd{4}$",fullName:"RESID"},{database:"Rfam",id:"Rf",homePage:"",linkOut:"http://www.sanger.ac.uk/cgi-bin/Rfam/getacc?$id",example:"RF00066",dataNodeType:"gene",species:"",priority:1,unknown:"Rf",regex:"RFd+",fullName:"RFAM"},{database:"RGD",id:"R",homePage:"http://rgd.mcw.edu/",linkOut:"http://rgd.mcw.edu/tools/genes/genes_view.cgi?id=$id",example:"2018",dataNodeType:"gene",species:"Rattus norvegicus",priority:1,unknown:"urn:miriam:rgd",regex:"^d{4,7}$",fullName:"Rat Genome Database"},{database:"Rhea",id:"Rh",homePage:"http://www.ebi.ac.uk/rhea/",linkOut:"http://www.ebi.ac.uk/rhea/reaction.xhtml?id=$id",example:"12345",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:rhea",regex:"^d{5}$",fullName:"Rhea"},{database:"Rice Ensembl Gene",id:"Os",homePage:"http://www.gramene.org/Oryza_sativa",linkOut:"http://www.gramene.org/Oryza_sativa/geneview?gene=$id",example:"LOC_Os04g54800",dataNodeType:"gene",species:"Oryza sativa",priority:1,unknown:"Os",regex:"",fullName:"Rice Ensembl Gene"},{database:"SGD",id:"D",homePage:"http://www.yeastgenome.org/",linkOut:"http://www.yeastgenome.org/cgi-bin/locus.fpl?dbid=$id",example:"S000028457",dataNodeType:"gene",species:"Saccharomyces cerevisiae",priority:1,unknown:"urn:miriam:sgd",regex:"^Sd+$",fullName:"SGD"},{database:"Small Molecule Pathway Database",id:"Sm",homePage:"http://www.smpdb.ca/pathways",linkOut:"http://pathman.smpdb.ca/pathways/$id/pathway",example:"SMP00001",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:smpdb",regex:"^SMPd{5}$",fullName:"Small Molecule Pathway Database"},{database:"SMART",id:"Sma",homePage:"http://smart.embl-heidelberg.de/",linkOut:"http://smart.embl-heidelberg.de/smart/do_annotation.pl?DOMAIN=$id",example:"SM00015",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:smart",regex:"^SMd{5}$",fullName:"SMART"},{database:"SPIKE",id:"Sk",homePage:"http://www.cs.tau.ac.il/~spike/",linkOut:"http://www.cs.tau.ac.il/~spike/maps/$id.html",example:"spike00001",dataNodeType:"interaction",species:"",priority:1,unknown:"urn:miriam:spike.map",regex:"^spiked{5}$",fullName:"SPIKE Map"},{database:"SPRINT",id:"Spr",homePage:"http://www.bioinf.manchester.ac.uk/dbbrowser/sprint/",linkOut:"http://www.bioinf.manchester.ac.uk/cgi-bin/dbbrowser/sprint/searchprintss.cgi?prints_accn=$id&display_opts=Prints&category=None&queryform=false&regexpr=off",example:"PR00001",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:sprint",regex:"^PRd{5}$",fullName:"SPRINT"},{database:"STRING",id:"Str",homePage:"http://string.embl.de/",linkOut:"http://string.embl.de/interactions/$id",example:"P53350",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:string",regex:"^([A-N,R-Z][0-9][A-Z][A-Z, 0-9][A-Z, 0-9][0-9])|([O,P,Q][0-9][A-Z, 0-9][A-Z, 0-9][A-Z, 0-9][0-9])|([0-9][A-Za-z0-9]{3})$",fullName:"STRING"},{database:"SubstrateDB",id:"Sdb",homePage:"http://substrate.burnham.org/",linkOut:"http://substrate.burnham.org/protein/annotation/$id/html",example:"1915",dataNodeType:"protein",species:"",priority:0,unknown:"urn:miriam:pmap.substratedb",regex:"^d+$",fullName:"SubstrateDB"},{database:"SubtiWiki",id:"Sw",homePage:"http://www.subtiwiki.uni-goettingen.de/wiki/index.php/Main_Page",linkOut:"http://www.subtiwiki.uni-goettingen.de/wiki/index.php/$id",example:"BSU29180",dataNodeType:"gene",species:"Bacillus subtilis",priority:1,unknown:"urn:miriam:subtiwiki",regex:"^BSUd{5}$",fullName:"SubtiWiki"},{database:"SUPFAM",id:"Sf",homePage:"http://supfam.org/SUPERFAMILY/",linkOut:"http://supfam.org/SUPERFAMILY/cgi-bin/scop.cgi?ipid=$id",example:"SSF57615",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:supfam",regex:"^w+$",fullName:"SUPFAM"},{database:"SWISS-MODEL",id:"Sw",homePage:"http://swissmodel.expasy.org/",linkOut:"http://swissmodel.expasy.org/repository/smr.php?sptr_ac=$id",example:"P23298",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:swiss-model",regex:"^w+$",fullName:"SWISS-MODEL"},{database:"Systems Biology Ontology",id:"Sbo",homePage:"http://www.ebi.ac.uk/sbo/",linkOut:"http://www.ebi.ac.uk/sbo/main/$id",example:"SBO:0000262",dataNodeType:"ontology",species:"",priority:1,unknown:"urn:miriam:biomodels.sbo",regex:"^SBO:d{7}$",fullName:"Systems Biology Ontology"},{database:"TAIR",id:"A",homePage:"http://arabidopsis.org/index.jsp",linkOut:"http://arabidopsis.org/servlets/TairObject?type=locus&name=$id",example:"AT1G01030",dataNodeType:"gene",species:"",priority:1,unknown:"urn:miriam:tair.locus",regex:"^AT[1-5]Gd{5}$",fullName:"TAIR Locus"},{database:"TIGR",id:"Ti",homePage:"http://www.jcvi.org/",linkOut:"",example:"12012.t00308",dataNodeType:"gene",species:"",priority:1,unknown:"Ti",regex:"",fullName:"TIGR"},{database:"TTD Drug",id:"Td",homePage:"http://bidd.nus.edu.sg/group/cjttd/TTD_HOME.asp",linkOut:"http://bidd.nus.edu.sg/group/cjttd/ZFTTDDRUG.asp?ID=$id",example:"DAP000773",dataNodeType:"metabolite",species:"",priority:1,unknown:"urn:miriam:ttd.drug",regex:"^DAPd+$",fullName:"TTD Drug"},{database:"TTD Target",id:"Tt",homePage:"http://bidd.nus.edu.sg/group/cjttd/TTD_HOME.asp",linkOut:"http://bidd.nus.edu.sg/group/cjttd/ZFTTDDetail.asp?ID=$id",example:"TTDS00056",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:ttd.target",regex:"^TTDSd+$",fullName:"TTD Target"},{database:"TubercuList",id:"Tb",homePage:"http://tuberculist.epfl.ch",linkOut:"http://tuberculist.epfl.ch/quicksearch.php?gene+name=$id",example:"Rv0064",dataNodeType:"gene",species:"Mycobacterium tuberculosis",priority:1,unknown:"Tb",regex:"Rvd{4}(A|B|c|.d)?",fullName:"TubercuList"},{database:"UCSC Genome Browser",id:"Uc",homePage:"http://genome.ucsc.edu/",linkOut:"http://genome.ucsc.edu/cgi-bin/hgTracks?position=$id",example:"uc001tyh.1",dataNodeType:"gene",species:"",priority:1,unknown:"Uc",regex:"ucd{3}[a-z]{3}.d",fullName:"UCSC Genome Browser"},{database:"UniGene",id:"U",homePage:"http://www.ncbi.nlm.nih.gov/sites/entrez?db=unigene",linkOut:"http://www.ncbi.nlm.nih.gov/UniGene/clust.cgi?UGID=1548618&SEARCH=$id",example:"Hs.553708",dataNodeType:"gene",species:"",priority:1,unknown:"U",regex:"[A-Z][a-z][a-z]?.d+",fullName:"UniGene"},{database:"Unipathway",id:"Up",homePage:"http://www.grenoble.prabi.fr/obiwarehouse/unipathway",linkOut:"http://www.grenoble.prabi.fr/obiwarehouse/unipathway/upa?upid=$id",example:"UPA00206",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:unipathway",regex:"^UPAd{5}$",fullName:"Unipathway"},{database:"Uniprot-TrEMBL",id:"S",homePage:"http://www.uniprot.org/",linkOut:"http://www.uniprot.org/uniprot/$id",example:"P62158",dataNodeType:"protein",species:"",priority:1,unknown:"urn:miriam:uniprot",regex:"^([A-N,R-Z][0-9][A-Z][A-Z, 0-9][A-Z, 0-9][0-9])|([O,P,Q][0-9][A-Z, 0-9][A-Z, 0-9][A-Z, 0-9][0-9])|($",fullName:"UniProtKB/TrEMBL"},{database:"Uniprot-SwissProt",id:"Sp",homePage:"http://www.uniprot.org/",linkOut:"http://www.uniprot.org/uniprot/$id",example:"CALM_HUMAN",dataNodeType:"protein",species:"",priority:1,unknown:"Sp",regex:"^[A-Z0-9]+_[A-Z]+$",fullName:"UniProtKB/Swiss-Prot"},{database:"Wheat gene names",id:"Wn",homePage:"http://wheat.pw.usda.gov/",linkOut:"http://wheat.pw.usda.gov/report?class=gene;name=$id",example:"5S-Rrna-D1_(Triticum)",dataNodeType:"gene",species:"Triticum aestivum",priority:1,unknown:"Wn",regex:"",fullName:"Wheat gene names"},{database:"Wheat gene refs",id:"Wr",homePage:"http://wheat.pw.usda.gov/",linkOut:"http://wheat.pw.usda.gov/cgi-bin/graingenes/report.cgi?class=reference&name=$id",example:"WGS-95-1333",dataNodeType:"probe",species:"Triticum aestivum",priority:0,unknown:"Wr",regex:"",fullName:"Wheat gene refs"},{database:"WikiGenes",id:"Wg",homePage:"http://www.wikigenes.org/",linkOut:"http://www.wikigenes.org/e/gene/e/$id.html",example:"7157",dataNodeType:"gene",species:"",priority:0,unknown:"Wg",regex:"",fullName:"WikiGenes"},{database:"WikiPathways",id:"Wp",homePage:"http://www.wikipathways.org/",linkOut:"http://www.wikipathways.org/index.php/Pathway:$id",example:"WP100",dataNodeType:"pathway",species:"",priority:1,unknown:"urn:miriam:wikipathways",regex:"WPd{1,5}",fullName:"WikiPathways"},{database:"Wikipedia",id:"Wi",homePage:"http://www.wikipedia.org",linkOut:"http://en.wikipedia.org/wiki/$id",example:"Acetate",dataNodeType:"metabolite",species:"",priority:0,unknown:"Wi",regex:"",fullName:"Wikipedia"},{database:"WormBase",id:"W",homePage:"http://www.wormbase.org/",linkOut:"http://www.wormbase.org/db/gene/gene?name=$id;class=Gene",example:"WBGene00000001",dataNodeType:"gene",species:"Caenorhabditis elegans",priority:1,unknown:"urn:miriam:wormbase",regex:"^WBGened{8}$",fullName:"WormBase"},{database:"ZFIN",id:"Z",homePage:"http://zfin.org",linkOut:"http://zfin.org/action/marker/view/$id",example:"ZDB-GENE-041118-11",dataNodeType:"gene",species:"Danio rerio",priority:1,unknown:"urn:miriam:zfin",regex:"ZDB-GENE-d+-d+",fullName:"ZFIN Gene"}],pathvisio.pathway.xRef=function(){function getData(species,database,id,callback){var databaseId=pathvisio.pathway.dataSources.filter(function(element){return element.database===database
})[0].id,url="../../remote-data-sources/php/bridgedb.php?species="+encodeURIComponent(species)+"&database="+encodeURIComponent(databaseId)+"&id="+encodeURIComponent(id);$.ajax({url:url,dataType:"text",success:function(data){callback(data)}})}function displayData(organism,node){console.log("organism"),console.log(organism),self.node=node,getData(organism,node.xRef.database,node.xRef.id,function(data){var parser=CSVParser.parse(data,!0," ",!1,!1,"."),parsed=DataGridRenderer.json(parser.dataGrid,parser.headerNames,parser.headerTypes,"	","\n"),xRefDataParsed=self.xRefDataParsed=JSON.parse(parsed),idsByDatabase=xRefDataParsed,feature={};idsByDatabase.ids=[];var features=[];xRefDataParsed.forEach(function(xRefForEach,index,array){feature.database=xRefForEach.database,feature.ids=[],0===features.filter(function(featureFilter){return featureFilter.database===xRefForEach.database}).length&&(array.filter(function(xRefFilter){return xRefFilter.database===xRefForEach.database}).forEach(function(element){feature.ids.push(element.id)}),features.push({database:xRefForEach.database,ids:feature.ids}))}),features.forEach(function(feature){try{var dataSource=pathvisio.pathway.dataSources.filter(function(dataSource){return dataSource.database.replace(/[^a-z0-9]/gi,"").toLowerCase()==feature.database.replace(/[^a-z0-9]/gi,"").toLowerCase()})[0];feature.dataSourceId=dataSource.id,feature.linkOut=dataSource.linkOut,feature.priority=dataSource.priority}catch(e){console.warn(e),console.warn('Error: No database found for external reference database "'+feature.database+'".')}}),features.sort(function(a,b){if(a.priority===b.priority){var x=a.database.toLowerCase(),y=b.database.toLowerCase();return y>x?-1:x>y?1:0}return b.priority-a.priority});var specifiedFeature=features.filter(function(element){return element.database==node.xRef.database})[0],currentFeatureIndex=features.indexOf(specifiedFeature),specifiedXRefId=specifiedFeature.ids.filter(function(element){return element==node.xRef.id})[0],currentXRefIdIndex=specifiedFeature.ids.indexOf(specifiedXRefId);features=pathvisio.helpers.moveArrayItem(features,currentFeatureIndex,0),specifiedFeature.ids=pathvisio.helpers.moveArrayItem(specifiedFeature.ids,currentXRefIdIndex,0);var detailsFrame=d3.select("#details-frame");detailsFrame.selectAll("*").remove();var detailsHeader=detailsFrame.append("header").attr("class","data-node-label"),detailsPullLeftSpan=detailsHeader.append("span").attr("class","pull-left"),detailsMoveSpan=detailsPullLeftSpan.append("span").attr("class","header-move");detailsMoveSpan.append("i").attr("class","icon-move").attr("style","color:#aaa");var detailsHeaderLabelSpan=detailsHeader.append("span").attr("style","font-size: 120%;").text(function(){return node.textLabel.text+" "}),detailsSearchSpan=detailsHeaderLabelSpan.append("span").attr("class","header-search").attr("title",function(){return"Search for pathways containing "+node.textLabel.text}),detailsSearchLink=detailsSearchSpan.append("a").attr("href",function(){return"http://wikipathways.org//index.php?title=Special:SearchPathways&doSearch=1&ids="+node.xRef.id+"&codes="+pathvisio.pathway.dataSources.filter(function(dataSource){return dataSource.database.replace(/[^a-z0-9]/gi,"").toLowerCase()==node.xRef.database.replace(/[^a-z0-9]/gi,"").toLowerCase()})[0].id+"&type=xref"});detailsSearchLink.append("i").attr("class","icon-search").attr("style","color:blue; font-size:50%");var detailsPullRightSpan=detailsHeader.append("span").attr("class","pull-right"),detailsCloseSpan=detailsPullRightSpan.append("span").attr("class","header-close").on("click",function(){detailsFrame.selectAll("*").remove(),detailsFrame[0][0].style.visibility="hidden"});detailsCloseSpan.append("i").attr("class","icon-remove").attr("style","color:#aaa; font-size:120%");var dataNodeTypeDiv=detailsHeader.append("div").attr("class","data-node-description");dataNodeTypeDiv.append("h2").text(node.dataNodeType);var detailsList=detailsFrame.append("ul").attr("class","data-node"),detailsListItems=detailsList.selectAll("li").data(features).enter().append("li");detailsListItems[0].forEach(function(detailsListItem){d3.select(detailsListItem).append("span").attr("class","feature-title").text(function(d){return d.database+": "});var linkOuts=d3.select(detailsListItem).selectAll("a").data(function(d){var featuresFilled=[];return d.ids.forEach(function(id){var linkOut=d.linkOut.replace("$id",id);featuresFilled.push({id:id,linkOut:linkOut})}),featuresFilled}).enter().append("a").attr("href",function(d){return d.linkOut});linkOuts.append("span").attr("class","feature-text").attr("style",function(d){return d.linkOut?"":"color: #696969;"}).text(function(d){return" "+d.id})}),detailsFrame[0][0].style.visibility="visible"})}return{getData:getData,displayData:displayData}}();